{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9efec848-2f2c-4deb-8544-62dacffc5e01",
   "metadata": {},
   "source": [
    "# GP2\n",
    "\n",
    "* **Project:** ADRD-SORL1-Biobanks\n",
    "* **Version:** Python/3.10\n",
    "* **Last Updated:** 10-Jul-2025\n",
    "\n",
    "## Notebook Overview\n",
    " Gene characterization, allele freqs, association analysis, burden analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fef5dfc8-2a8d-46ad-8884-78e971eabeb5",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Workspace Resources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b48de81a-3942-4259-b81b-6c1fabbf1b58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# mount resources\n",
    "! wb resource mount --id=gp2_tier2_eu_release9_18122024"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b01808c7-d16d-48c5-b772-85066d19d2a3",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "72752ad3-8a1d-4f08-917a-1769a68abc1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import the necessary packages \n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import sys\n",
    "import subprocess\n",
    "import statsmodels.api as sm\n",
    "import scipy\n",
    "from scipy import stats\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from IPython.display import display\n",
    "import pathlib"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "042624f9-51d8-4fa0-b362-2a4c3f478ba2",
   "metadata": {},
   "source": [
    "## Package Installs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1935c7d6-0267-42e3-81e1-4550abdc1fb0",
   "metadata": {},
   "source": [
    "### PLINK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bef5b8ca-2fc9-4100-9b6c-e4486a04c984",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir -p ~/tools\n",
    "cd ~/tools\n",
    "\n",
    "if test -e /home/jupyter/tools/plink; then\n",
    "echo \"Plink1.9 is already installed in /home/jupyter/tools/\"\n",
    "\n",
    "else\n",
    "echo -e \"Downloading plink \\n    -------\"\n",
    "wget -N http://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20190304.zip \n",
    "unzip -o plink_linux_x86_64_20190304.zip\n",
    "echo -e \"\\n plink downloaded and unzipped in /home/jupyter/tools \\n \"\n",
    "\n",
    "fi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "567111cb-0127-442d-a6cb-49e19d451872",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir -p ~/tools\n",
    "cd ~/tools\n",
    "\n",
    "if test -e /home/jupyter/tools/plink2; then\n",
    "echo \"Plink2 is already installed in /home/jupyter/tools/\"\n",
    "\n",
    "else\n",
    "echo -e \"Downloading plink2 \\n    -------\"\n",
    "wget -N https://s3.amazonaws.com/plink2-assets/alpha5/plink2_linux_x86_64_20240820.zip\n",
    "unzip -o plink2_linux_x86_64_20240820.zip\n",
    "echo -e \"\\n plink2 downloaded and unzipped in /home/jupyter/tools \\n \"\n",
    "\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6e1438e-6361-480e-8fca-0ce9a0162f15",
   "metadata": {},
   "source": [
    "### ANNOVAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "910f2906-523e-4dce-b286-415d2ee70207",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Install ANNOVAR:\n",
    "# https://www.openbioinformatics.org/annovar/annovar_download_form.php\n",
    "\n",
    "if test -e /home/jupyter/tools/annovar; then\n",
    "\n",
    "echo \"annovar is already installed in /home/jupyter/tools/\"\n",
    "else\n",
    "echo \"annovar is not installed\"\n",
    "cd /home/jupyter/tools/\n",
    "\n",
    "wget http://www.openbioinformatics.org/annovar/download/0wgxR2rIVP/annovar.latest.tar.gz\n",
    "\n",
    "tar xvfz annovar.latest.tar.gz\n",
    "\n",
    "fi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1f28b14a-6194-4d1f-b134-c752e300577d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%capture\n",
    "%%bash\n",
    "\n",
    "# Install ANNOVAR: Download resources for annotation\n",
    "\n",
    "cd /home/jupyter/tools/annovar/\n",
    "\n",
    "perl annotate_variation.pl -buildver hg38 -downdb -webfrom annovar refGene humandb/\n",
    "perl annotate_variation.pl -buildver hg38 -downdb -webfrom annovar dbnsfp47a humandb/\n",
    "#perl annotate_variation.pl -buildver hg38 -downdb -webfrom annovar clinvar_20240611 humandb/\n",
    "#perl annotate_variation.pl -buildver hg38 -downdb -webfrom annovar gnomad41_exome humandb/\n",
    "#perl annotate_variation.pl -buildver hg38 -downdb cytoBand humandb/\n",
    "#perl annotate_variation.pl -buildver hg38 -downdb -webfrom annovar ensGene humandb/\n",
    "#perl annotate_variation.pl -buildver hg38 -downdb -webfrom annovar exac03 humandb/ \n",
    "#perl annotate_variation.pl -buildver hg38 -downdb -webfrom annovar avsnp147 humandb/ \n",
    "#perl annotate_variation.pl -buildver hg38 -downdb -webfrom annovar ljb26_all humandb/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93ec1340-20c3-49e6-b3bb-6c518d4d93da",
   "metadata": {},
   "source": [
    "### RVTests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "505391c6-8c16-439e-80c9-27ea6e00d5db",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#Install RVTESTS: Option 1 (~15min)\n",
    "if test -e /home/jupyter/tools/rvtests; then\n",
    "\n",
    "echo \"rvtests is already installed\"\n",
    "else\n",
    "echo \"rvtests is not installed\"\n",
    "\n",
    "mkdir /home/jupyter/tools/rvtests\n",
    "cd /home/jupyter/tools/rvtests\n",
    "\n",
    "wget https://github.com/zhanxw/rvtests/releases/download/v2.1.0/rvtests_linux64.tar.gz \n",
    "\n",
    "tar -zxvf rvtests_linux64.tar.gz\n",
    "fi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b9e9a6b7-8523-4687-9d20-e002769a1545",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# chmod to make sure you have permission to run the program\n",
    "! chmod u+x /home/jupyter/tools/plink\n",
    "! chmod u+x /home/jupyter/tools/plink2\n",
    "! chmod 777 /home/jupyter/tools/rvtests/executable/rvtest"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1acec66-0ca2-40a1-b401-eca4b023ef68",
   "metadata": {},
   "source": [
    "## Workbook Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bd126db-84d5-4f5a-8d90-da2674c499f8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Edit ancestry label and run this cell and all below for remaining ancestries\n",
    "ancestry = \"EUR\"\n",
    "CHROM = \"chr11\"\n",
    "GENE = \"SORL1\"\n",
    "WORK_DIR =  f\"/home/jupyter/{GENE}_results/{ancestry}_R9\"\n",
    "DATA_DIR = f\"/home/jupyter/workspace\"\n",
    "\n",
    "! mkdir -p {WORK_DIR}\n",
    "%cd {WORK_DIR}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4d13a1f-bda9-4e6a-86d9-7c3ad4e71175",
   "metadata": {},
   "source": [
    "## Make Covariate File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97292067-c7ed-4be6-87a2-6d1b4cf4de4f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Let's load the master key\n",
    "key = pd.read_csv(f\"{DATA_DIR}/gp2_tier2_eu_release9_18122024/clinical_data/master_key_release9_final_vwb.csv\")\n",
    "print(key.shape)\n",
    "key = key[[\"GP2ID\", \"baseline_GP2_phenotype\", \"biological_sex_for_qc\", \"age_at_sample_collection\", \"age_of_onset\", \"age_at_diagnosis\",\"age_at_last_follow_up\", \"race_for_qc\", \"nba_label\"]]\n",
    "key.rename(columns = {\"GP2ID\": \"IID\",\n",
    "                      \"baseline_GP2_phenotype\":\"phenotype\",\n",
    "                                     \"biological_sex_for_qc\":\"SEX\", \n",
    "                                     \"age_at_sample_collection\":\"AGE\",\n",
    "                                     \"race_for_qc\":\"RACE\",\n",
    "                                     \"age_at_diagnosis\":\"AAD\",\n",
    "                                     \"age_at_last_follow_up\":\"AAFU\",\n",
    "                                     \"age_of_onset\":\"AAO\"}, inplace = True)\n",
    "\n",
    "## Subset to keep ancestry of interest \n",
    "ancestry_key = key[key[\"nba_label\"]==ancestry].copy()\n",
    "ancestry_key.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0b5f2fe-fd2f-4714-8efd-ce2cf0d004dc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load information about related individuals \n",
    "related_df = pd.read_csv(f\"{DATA_DIR}/gp2_tier2_eu_release9_18122024/meta_data/related_samples/{ancestry}_release9_vwb.related\")\n",
    "print(related_df.shape)\n",
    "\n",
    "# Make a list of just one set of related people\n",
    "related_list = list(related_df[\"IID1\"].str.rstrip(\"_s1\"))\n",
    "related_list\n",
    "# Check value counts of related and remove only one related individual\n",
    "ancestry_key = ancestry_key[~ancestry_key[\"IID\"].isin(related_list)]\n",
    "\n",
    "# remove related individuals\n",
    "print(f\"Removing {len(related_list)} individuals.\")\n",
    "\n",
    "# Check size\n",
    "print(ancestry_key.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f673725-3419-4c5a-ab17-b42cf972c2f1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Convert phenotype to binary (1/2)\n",
    "## Assign conditions so case=2 and controls=1, and -9 otherwise (matching PLINK convention)\n",
    "    # PD = 2; control = 1\n",
    "pheno_mapping = {\"PD\": 2, \"Control\": 1}\n",
    "ancestry_key[\"PHENO\"] = ancestry_key[\"phenotype\"].map(pheno_mapping).astype(\"Int64\")\n",
    "# Check value counts of pheno\n",
    "ancestry_key[\"PHENO\"].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "460df1b6-df90-41db-adf6-98595626c92d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Convert phenotype to binary (1/2)\n",
    "\n",
    "# Female = 2; Male = 1\n",
    "sex_mapping = {\"Female\": 2, \"Male\": 1}\n",
    "ancestry_key[\"SEX\"] = ancestry_key[\"SEX\"].map(sex_mapping).astype(\"Int64\")\n",
    "\n",
    "# Check value counts of SEX\n",
    "ancestry_key[\"SEX\"].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb74106d-ad04-4a8e-9c02-71a51de00e85",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# only include samples that are in the PLINK files\n",
    "\n",
    "genotyped_samples = pd.read_csv(f\"{DATA_DIR}/gp2_tier2_eu_release9_18122024/imputed_genotypes/{ancestry}/{CHROM}_{ancestry}_release9_vwb.psam\", sep = \"\\t\")[\"#IID\"]\n",
    "genotyped_samples = list(genotyped_samples)\n",
    "\n",
    "print(f\"\"\"Dropping {(~ancestry_key[\"IID\"].isin(genotyped_samples)).sum()} samples without genotyping\"\"\")\n",
    "ancestry_key = ancestry_key[ancestry_key[\"IID\"].isin(genotyped_samples)]\n",
    "ancestry_key.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b78d9ad-8c1c-484c-9345-5ff7017c980c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Get the PCs\n",
    "pcs = pd.read_csv(f\"{DATA_DIR}/gp2_tier2_eu_release9_18122024/raw_genotypes/{ancestry}/{ancestry}_release9_vwb.eigenvec\", sep = \"\\t\")\n",
    "selected_columns = [\"IID\", \"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\", \"PC6\", \"PC7\", \"PC8\", \"PC9\",\"PC10\"]\n",
    "pcs = pd.DataFrame(data=pcs.iloc[:, 1:12].values, columns=selected_columns)\n",
    "pcs[\"IID\"] = pcs[\"IID\"].str.replace(\"_s1\", \"\", case= True)\n",
    "\n",
    "# Reset the index to remove any potential issues\n",
    "pcs = pcs.reset_index(drop=True)\n",
    "\n",
    "# Display the resulting DataFrame\n",
    "print(pcs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49312681-7b8e-4013-99ba-6e70ac1a5109",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Make covariate file\n",
    "df = pd.merge(ancestry_key, pcs, on=\"IID\", how= \"inner\")\n",
    "\n",
    "## Drop lines with missing pheno\n",
    "df = df[df[\"PHENO\"].notna()]\n",
    "\n",
    "print(f\"\"\"Dropping {df[\"PHENO\"].isna().sum()} with missing pheno \"\"\")\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93942292-a8fd-47ed-a266-bba5ee57901c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "ded6c019-00f8-4c80-8988-84ecc63d9f68",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df[\"FID\"] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04869a97-7611-4dc6-8a16-9a48198508e3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Clean up and keep columns we need \n",
    "final_df = df[[\"FID\",\"IID\", \"SEX\", \"AGE\", \"AAO\", \"PHENO\", \"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\", \"PC6\", \"PC7\", \"PC8\", \"PC9\",\"PC10\"]].copy()\n",
    "final_df.groupby([\"PHENO\"])[\"SEX\"].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af32c646-73e9-49cc-b2b0-16838cdece44",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Make file of sample IDs to keep \n",
    "samples_toKeep = final_df[[\"FID\", \"IID\"]].copy()\n",
    "\n",
    "samples_toKeep.to_csv(f\"{WORK_DIR}/{ancestry}.samplestoKeep\", sep = \"\\t\", index=False, header=None)\n",
    "samples_toKeep.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b8d7433-b07f-43e1-8b4d-85bc18a31840",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make file of just cases to keep \n",
    "cases_toKeep = final_df[final_df[\"PHENO\"]==2][[\"FID\", \"IID\"]]\n",
    "cases_toKeep.to_csv(f\"{WORK_DIR}/{ancestry}.cases_toKeep\", sep = \"\\t\", index = False, header=None)\n",
    "cases_toKeep.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "9cc324b7-364a-4a70-8827-6e25b47ba5db",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Save your covariate file\n",
    "final_df.to_csv(f\"{WORK_DIR}/{ancestry}_covariate_file.txt\", sep = \"\\t\", index=False, header=True, na_rep=\"NA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "950e34c1-0695-4e51-b915-d0655e66df3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b59dd496-1904-4f90-80f6-2590a1ee5080",
   "metadata": {},
   "source": [
    "### Get cohort summary stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef85aa28-0ce0-4c53-ae1a-e045ea335e58",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### \n",
    "print(f\"\"\"\n",
    "Number of {ancestry} PD cases: {(final_df[\"PHENO\"] == 2).sum()}\n",
    "Number of {ancestry} Controls: {(final_df[\"PHENO\"] == 1).sum()}\n",
    "\n",
    "{final_df.groupby([\"PHENO\"])[\"SEX\"].value_counts(sort=True)} \n",
    "\n",
    "PD age range: {final_df[final_df[\"PHENO\"] == 2][\"AGE\"].mean().round(2)} +/- {round(final_df[final_df[\"PHENO\"] == 2][\"AGE\"].std(),2)}\n",
    "Controls age range: {final_df[final_df[\"PHENO\"] == 1][\"AGE\"].mean().round(2)} +/- {round(final_df[final_df[\"PHENO\"] == 1][\"AGE\"].std(),2)}\n",
    "\n",
    "nan_counts:\\n{final_df.isna().sum()}\n",
    "\"\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3d87b40-6999-421f-8264-9e65ea30d815",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# extract region of interest from file\n",
    "! /home/jupyter/tools/plink2 \\\n",
    "--pfile {DATA_DIR}/gp2_tier2_eu_release9_18122024/imputed_genotypes/{ancestry}/{CHROM}_{ancestry}_release9_vwb \\\n",
    "--chr 11 \\\n",
    "--from-bp 121352314 \\\n",
    "--to-bp 121733763 \\\n",
    "--keep {WORK_DIR}/{ancestry}.samplestoKeep \\\n",
    "--make-pgen erase-dosage \\\n",
    "--out {WORK_DIR}/{CHROM}_{ancestry}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e6c1ec0-bc8a-4570-969e-67159227c9e3",
   "metadata": {},
   "source": [
    "## Annotation using ANNOVAR\n",
    "- *SORL1* from NCBI gene, with gene region flanked by 100kb on both sides\n",
    "- hg38 (chr11:121452314-121633763) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03a0d87b-bf18-4355-b83b-93d6e2fe7ffb",
   "metadata": {},
   "source": [
    "### Annotate Variants for All Cases and Controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "8daf70bc-2e51-4eb5-9505-9eaa41cabadf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set file prefix for vcf with all cases and controls\n",
    "PLINK_PREFIX = f\"{CHROM}_{GENE}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0582c94-7616-4d1d-9b15-297535046cae",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## extract region of interest using plink and Convert the files to VCF format\n",
    "! /home/jupyter/tools/plink2 \\\n",
    "--pfile {WORK_DIR}/{CHROM}_{ancestry} \\\n",
    "--chr 11 \\\n",
    "--from-bp 121452314 \\\n",
    "--to-bp 121633763 \\\n",
    "--keep {WORK_DIR}/{ancestry}.samplestoKeep \\\n",
    "--recode vcf id-paste=iid \\\n",
    "--mac 2 \\\n",
    "--out {WORK_DIR}/{PLINK_PREFIX}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "adc347fa-db2f-4676-9cf1-8deda7d37d10",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### Bgzip and Tabix (zip and index the file)\n",
    "! bgzip -f {WORK_DIR}/{PLINK_PREFIX}.vcf\n",
    "! tabix -f -p vcf {WORK_DIR}/{PLINK_PREFIX}.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3f4488e-2d9f-44e6-b306-eb3b33820e45",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# annotating just one sample id to make annotation run faster\n",
    "samples = pd.read_csv(f\"{WORK_DIR}/{ancestry}.samplestoKeep\", sep = \"\\t\", header=None)\n",
    "sample_id = samples.iloc[0,1]\n",
    "filename = f\"{WORK_DIR}/{PLINK_PREFIX}.vcf.gz\"\n",
    "sample_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ef734f0-86ab-4dcd-a315-4903e82ac918",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### export only one sample id\n",
    "! /home/jupyter/tools/plink2 \\\n",
    "--vcf {filename} \\\n",
    "--indv {sample_id} \\\n",
    "--export vcf bgz \\\n",
    "--out {WORK_DIR}/{ancestry}_{GENE}_{sample_id}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "f76ced69-3794-44ad-b2c2-f461f6e8a599",
   "metadata": {},
   "outputs": [],
   "source": [
    "PREFIX = f\"{ancestry}_{GENE}_{sample_id}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcb6a31a-6813-4bce-9d85-b140476a571e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s $PREFIX $GENE\n",
    "## annotate using ANNOVAR\n",
    "\n",
    "for i in $1\n",
    "do\n",
    "    /home/jupyter/tools/annovar/table_annovar.pl ${i}.vcf.gz /home/jupyter/tools/annovar/humandb/ --buildver hg38 --out ${i}.annovar --remove --protocol refGene,dbnsfp47a --operation g,f --otherinfo --polish --nastring . --vcfinput\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10c2430c-031f-4ad0-b596-42f0cf02786c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# check file was made correctly\n",
    "! ls $PREFIX*multianno*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "e151656b-6bc3-4890-9e8d-b3b5725a2bd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename annovar multianno to remove subject id\n",
    "! mv -f {PREFIX}.annovar.hg38_multianno.txt {ancestry}_{GENE}.annovar.hg38_multianno.txt\n",
    "! mv -f {PREFIX}.annovar.hg38_multianno.vcf {ancestry}_{GENE}.annovar.hg38_multianno.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14af8d89-210c-4319-aab6-6c39d29bade0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Keep only columns of interest to make a manageable file \n",
    "! cut -f 1-10,102 {WORK_DIR}/{ancestry}_{GENE}.annovar.hg38_multianno.txt > {WORK_DIR}/{ancestry}_{GENE}.annovar.hg38_multianno.subset.txt\n",
    "! head {WORK_DIR}/{ancestry}_{GENE}.annovar.hg38_multianno.subset.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "715ebb38-730e-4f03-a422-d44430871cbd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Read in subsetted ANNOVAR multianno file\n",
    "gene = pd.read_csv(f\"{WORK_DIR}/{ancestry}_{GENE}.annovar.hg38_multianno.subset.txt\", sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41f37ff5-5a65-4b03-bd59-62283676f0cd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gene[\"Gene.refGene\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "b9196fe8-1274-4944-9605-66a2f2ce57a1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Keep only annotated ITSN1 variants\n",
    "gene_subset = gene[gene[\"Gene.refGene\"].isin([\"SORL1\", \"SC5D,SORL1\", \"SORL1,MIR100HG\"])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "4795b604-673f-487d-9bf5-32a03405dea9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gene_subset[[\"Func.refGene\", \"ExonicFunc.refGene\"]].value_counts().to_csv(f\"{WORK_DIR}/{ancestry}_{GENE}_variant_counts.txt\", index=True, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6be484e7-1037-4d45-902c-d2e0726a08f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "! cat {WORK_DIR}/{ancestry}_{GENE}_variant_counts.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71a81b96-587c-4923-98c9-990d9eef0d9b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gene_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "cf5f68c6-950f-4f7a-8bec-1a69cef4c094",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter intronic\n",
    "intronic = gene_subset[(gene_subset[\"Func.refGene\"] == \"intronic\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "d8221175-4d7a-4a06-91fe-bbacc6b0616c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter UTR3\n",
    "utr3 = gene_subset[(gene_subset[\"Func.refGene\"] == \"UTR3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "f899c8bb-7c99-4a3a-9bb9-81a0a412c161",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter UTR5\n",
    "utr5 = gene_subset[(gene_subset[\"Func.refGene\"] == \"UTR5\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "4f6b2be9-fcdb-429f-9797-b0697fd765f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter exonic and synonymous variants\n",
    "coding_synonymous = gene_subset[(gene_subset[\"Func.refGene\"] == \"exonic\") & (gene_subset[\"ExonicFunc.refGene\"] == \"synonymous SNV\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "02c915e7-c168-495f-970a-2b0bf14ca001",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter exonic and non-synonymous variants\n",
    "coding_nonsynonymous = gene_subset[(gene_subset[\"Func.refGene\"]== \"exonic\") & (gene_subset[\"ExonicFunc.refGene\"]==\"nonsynonymous SNV\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "7dd88272-b0cc-4ecd-8f71-11e893d72816",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter splicing and stopgain\n",
    "splicing = gene_subset[(gene_subset[\"Func.refGene\"] == \"splicing\")]\n",
    "stopgain = gene_subset[(gene_subset[\"Func.refGene\"] == \"exonic\") & (gene_subset[\"ExonicFunc.refGene\"] == \"stopgain\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "2e206726-dd3f-4c5c-9104-67a431d9e3f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine splicing and stopgain variants\n",
    "all_nonsynonymous = pd.concat([coding_nonsynonymous, splicing, stopgain], axis=0, ignore_index=True)\n",
    "all_nonsynonymous\n",
    "\n",
    "# ensure splicing and stopgain labels are in annotations\n",
    "all_nonsynonymous[\"ExonicFunc.refGene\"] = all_nonsynonymous.apply(lambda row: row[\"Func.refGene\"] if row[\"ExonicFunc.refGene\"] == \".\" else row[\"ExonicFunc.refGene\"], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "66b184ba-008f-4cab-b941-279d00855461",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count total variants and subtypes\n",
    "total_variants = len(gene_subset)\n",
    "total_intronic = len(intronic)\n",
    "total_utr3 = len(utr3)\n",
    "total_utr5 = len(utr5)\n",
    "total_exonic_syn = len(coding_synonymous)\n",
    "total_exonic_nonsyn = len(coding_nonsynonymous)\n",
    "total_splicing = len(splicing)\n",
    "total_stopgain = len(stopgain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8dc9779-aa82-4c53-8e8d-66c76c972231",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Print summary of counts\n",
    "print(f\"Total Variants in {GENE}: {total_variants:,}\")\n",
    "print(f\"    - Intronic: {total_intronic:,}\")\n",
    "print(f\"    - UTR3: {total_utr3:,}\")\n",
    "print(f\"    - UTR5: {total_utr5:,}\")\n",
    "print(f\"    - Exonic Synonymous: {total_exonic_syn:,}\")\n",
    "print(f\"    - Exonic Non-synonymous: {total_exonic_nonsyn:,}\")\n",
    "print(f\"    - Splicing: {total_splicing:,}\")\n",
    "print(f\"    - Stopgain: {total_stopgain:,}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c837a628-1d9e-4f06-9a66-73b15f8105ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_nonsynonymous[\"SNP\"] = \"chr11:\" + all_nonsynonymous[\"Start\"].astype(str) + \":\" + all_nonsynonymous[\"Ref\"] + \":\" + all_nonsynonymous[\"Alt\"]\n",
    "all_nonsynonymous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "83ba9117-7763-46ca-b3ab-1d88cb989482",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure CADD values are numeric\n",
    "all_nonsynonymous[\"CADD_phred\"] = pd.to_numeric(all_nonsynonymous[\"CADD_phred\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74d65660-d0c0-4be0-b1bf-f6d91973f1c9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Save IDS to PLINK format \n",
    "variants_toKeep = all_nonsynonymous[\"SNP\"].copy()\n",
    "variants_toKeep.to_csv(f\"{WORK_DIR}/{ancestry}_{GENE}.all_coding_nonsyn.variantstoKeep.txt\", sep=\"\\t\", index=False, header=False)\n",
    "variants_toKeep.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "12ee560c-b7e5-4872-a033-5fbcf90610be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save annotation file for adding annotations later\n",
    "all_nonsynonymous.to_csv(f\"{WORK_DIR}/{ancestry}_{GENE}_coding_nonsyn_variant_annotations.txt\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bf63570-af19-4525-a20a-228230cd5f42",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## check to make sure file was created and saved\n",
    "! ls {WORK_DIR}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b51e214a-9011-4e91-a2bc-cf259b07dcbe",
   "metadata": {},
   "source": [
    "## Burden Analyses using RVTests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7aa5d08d-2313-41cc-b8d9-f74d8d940cb8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# get hg38 refFlat file from ucsc\n",
    "! wget -nc -O /home/jupyter/refFlat.txt.gz https://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/refFlat.txt.gz\n",
    "! gunzip -q -f /home/jupyter/refFlat.txt.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "a0dfac37-73b7-4345-8629-c8e8dbf765d8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# make a pheno file for plink input\n",
    "! cut -f 1,2,6 {WORK_DIR}/{ancestry}_covariate_file.txt > {WORK_DIR}/{ancestry}_pheno.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4793684-c8cd-4410-9553-4779a663cf8c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Convert the files from Plink 2.0 to Plink 1.9 format \n",
    "! /home/jupyter/tools/plink2 --pfile {WORK_DIR}/{CHROM}_{ancestry} \\\n",
    "--make-bed \\\n",
    "--pheno {ancestry}_pheno.txt \\\n",
    "--no-psam-pheno \\\n",
    "--pheno-name PHENO \\\n",
    "--real-ref-alleles \\\n",
    "--mac 2 \\\n",
    "--max-alleles 2 \\\n",
    "--out {WORK_DIR}/{CHROM}_{ancestry}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d38194b-c0d4-4643-9aea-9ff25caf4a69",
   "metadata": {},
   "outputs": [],
   "source": [
    "## extract variants\n",
    "! /home/jupyter/tools/plink \\\n",
    "--bfile {WORK_DIR}/{CHROM}_{ancestry} \\\n",
    "--keep {WORK_DIR}/{ancestry}.samplestoKeep \\\n",
    "--extract {WORK_DIR}/{ancestry}_{GENE}.all_coding_nonsyn.variantstoKeep.txt \\\n",
    "--real-ref-alleles \\\n",
    "--recode vcf-iid \\\n",
    "--out {WORK_DIR}/{ancestry}_{GENE}.coding_nonsyn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "0911c837-341a-468e-9e89-9bcdbe677aa6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### Bgzip and Tabix (zip and index the file)\n",
    "! bgzip -f {WORK_DIR}/{ancestry}_{GENE}.coding_nonsyn.vcf\n",
    "! tabix -f -p vcf {WORK_DIR}/{ancestry}_{GENE}.coding_nonsyn.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d1eaaea-4e28-4eb7-accf-4507e8963d6d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# create cov file that matches rvtest pheno file header\n",
    "rv_df = pd.read_csv(f\"{WORK_DIR}/{ancestry}_covariate_file.txt\", sep=\"\\t\")\n",
    "rv_df.columns = rv_df.columns.str.lower()\n",
    "rv_df[\"fatid\"] = 0\n",
    "rv_df[\"matid\"] = 0\n",
    "rv_df = rv_df[[\"fid\", \"iid\", \"fatid\", \"matid\", \"sex\", \"pheno\", \"age\", \"pc1\", \"pc2\", \"pc3\", \"pc4\", \"pc5\",\"pc6\",\"pc7\",\"pc8\",\"pc9\",\"pc10\"]]\n",
    "rv_df.to_csv(f\"{WORK_DIR}/rvtests_covariate_file.txt\", sep=\"\\t\", index=False, header=True)\n",
    "rv_df[\"age\"].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea3b54b6-f5b8-4744-8ced-298b1896afb7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## RVtests with covariates \n",
    "! /home/jupyter/tools/rvtests/executable/rvtest --noweb --hide-covar \\\n",
    "--out {WORK_DIR}/{ancestry}_{GENE}.burden.coding_nonsyn \\\n",
    "--kernel skat,skato \\\n",
    "--inVcf {WORK_DIR}/{ancestry}_{GENE}.coding_nonsyn.vcf.gz \\\n",
    "--pheno {WORK_DIR}/rvtests_covariate_file.txt \\\n",
    "--pheno-name pheno \\\n",
    "--gene {GENE} \\\n",
    "--geneFile /home/jupyter/refFlat.txt \\\n",
    "--covar {WORK_DIR}/rvtests_covariate_file.txt \\\n",
    "--covar-name sex,age,pc1,pc2,pc3,pc4,pc5,pc6,pc7,pc8,pc9,pc10\n",
    "\n",
    "# --out : Name of output \n",
    "# --burden cmc --kernel skato: tests to run \n",
    "# --inVcf : VCF file \n",
    "# --gene: gene name (if only looking at one or a few)\n",
    "# --geneFile refFlat.txt\n",
    "# --pheno :  covar file\n",
    "# --mpheno : # column that has phenotype information\n",
    "# --pheno-name : column name with phenotype in file\n",
    "# --covar : covar file\n",
    "# --freqUpper : optional, MAF cut-off\n",
    "# --covar-name : covariates, listed by column name, separated by commas (no spaces between commas)\n",
    "## 1=controls; 2=cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcfbe35f-a0cd-4f39-9378-40b59fabeb34",
   "metadata": {},
   "outputs": [],
   "source": [
    "## look at results \n",
    "! cat {WORK_DIR}/{ancestry}_{GENE}.burden.coding_nonsyn.Skat.assoc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abfcbc20-5bdd-4ef1-a8c1-f9b7ca590b63",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "! cat {WORK_DIR}/{ancestry}_{GENE}.burden.coding_nonsyn.SkatO.assoc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b415ed7e-ff28-413a-856e-d9da50bd98c5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## check to make sure file was created and saved\n",
    "! ls {WORK_DIR}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5ce711b-9eeb-4167-9818-a15a9934c2ac",
   "metadata": {},
   "source": [
    "## Case/Control Frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "841716a0-e9ec-47f1-be7a-c5b7f93933cb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# create new phenotype file (trying to fix issue with plink not reading phenotype column as C/C)\n",
    "#! cut -f 1,2,6 {WORK_DIR}/{ancestry}_covariate_file.txt > {WORK_DIR}/{ancestry}_pheno_file.txt\n",
    "#! head {WORK_DIR}/{ancestry}_pheno_file.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b040914a-5eee-465c-a9e9-89d534e68fe1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## run association test for F_A and F_U\n",
    "! /home/jupyter/tools/plink \\\n",
    "--bfile {WORK_DIR}/{CHROM}_{ancestry} \\\n",
    "--extract {WORK_DIR}/{ancestry}_{GENE}.all_coding_nonsyn.variantstoKeep.txt \\\n",
    "--keep {WORK_DIR}/{ancestry}.samplestoKeep \\\n",
    "--pheno {WORK_DIR}/{ancestry}_pheno.txt \\\n",
    "--mpheno 1 \\\n",
    "--assoc \\\n",
    "--allow-no-sex \\\n",
    "--ci 0.95 \\\n",
    "--out {WORK_DIR}/{ancestry}_{GENE}.coding_nonsyn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3afdbe81-78f9-4fc4-a464-0f8c901bae59",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# run logistic regression for pvals and odds ratios\n",
    "! /home/jupyter/tools/plink2 \\\n",
    "--vcf {WORK_DIR}/{ancestry}_{GENE}.coding_nonsyn.vcf.gz \\\n",
    "--adjust \\\n",
    "--keep {WORK_DIR}/{ancestry}.samplestoKeep \\\n",
    "--pheno {WORK_DIR}/{ancestry}_pheno.txt \\\n",
    "--ci 0.95 \\\n",
    "--covar {WORK_DIR}/{ancestry}_covariate_file.txt \\\n",
    "--covar-name SEX,AGE,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10 \\\n",
    "--covar-variance-standardize \\\n",
    "--glm hide-covar omit-ref firth-fallback cols=+a1freq,+a1freqcc,+a1count,+totallele,+a1countcc,+totallelecc,+gcountcc,+err \\\n",
    "--out {WORK_DIR}/{ancestry}_{GENE}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c5d31a1-af32-4000-8dd9-ad864dfec20a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# make bfiles with correct phenotypes\n",
    "! /home/jupyter/tools/plink2 \\\n",
    "--pfile {CHROM}_{ancestry} \\\n",
    "--make-bed \\\n",
    "--max-alleles 2 \\\n",
    "--pheno {WORK_DIR}/{ancestry}_pheno.txt \\\n",
    "--fa /home/jupyter/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna \\\n",
    "--pheno-name PHENO \\\n",
    "--out {WORK_DIR}/{CHROM}_{ancestry}_wPheno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f755e89f-e3b7-49ee-ae4b-232f93c7c724",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate recode file\n",
    "! /home/jupyter/tools/plink \\\n",
    "--bfile {WORK_DIR}/{CHROM}_{ancestry}_wPheno \\\n",
    "--extract {WORK_DIR}/{ancestry}_{GENE}.all_coding_nonsyn.variantstoKeep.txt \\\n",
    "--keep {WORK_DIR}/{ancestry}.samplestoKeep \\\n",
    "--recode A \\\n",
    "--out {WORK_DIR}/{ancestry}_{GENE}.coding_nonsyn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ca3a767-44d0-421a-9181-e443d19be6ec",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "recode = pd.read_csv(f\"{WORK_DIR}/{ancestry}_{GENE}.coding_nonsyn.raw\", sep=\"\\s+\")\n",
    "recode.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63236fbb-68fd-4f29-b5b4-92d5d3cfa153",
   "metadata": {},
   "outputs": [],
   "source": [
    "### merge output files\n",
    "\n",
    "ASSOC_FILE = f\"{WORK_DIR}/{ancestry}_{GENE}.coding_nonsyn.assoc\"\n",
    "RECODE_FILE = f\"{WORK_DIR}/{ancestry}_{GENE}.coding_nonsyn.raw\"\n",
    "GLM_HYBRID_FILE = f\"{WORK_DIR}/{ancestry}_{GENE}.PHENO.glm.logistic.hybrid\"\n",
    "GLM_ADJUST_FILE = f\"{WORK_DIR}/{ancestry}_{GENE}.PHENO.glm.logistic.hybrid.adjusted\"\n",
    "\n",
    "log_hybrid = pd.read_csv(GLM_HYBRID_FILE, sep=\"\\s+\", usecols=[\"ID\", \"A1\", \"A1_FREQ\", \"OBS_CT\", \"P\", \"OR\", \"LOG(OR)_SE\", \"L95\", \"U95\"])\n",
    "log_hybrid.rename(columns={\"ID\":\"SNP\"}, inplace=True)\n",
    "\n",
    "assoc_adjusted = pd.read_csv(GLM_ADJUST_FILE,  sep=\"\\s+\", usecols=[\"ID\", \"BONF\"])\n",
    "assoc_adjusted.rename(columns={\"ID\":\"SNP\"}, inplace=True)\n",
    "\n",
    "assoc = pd.read_csv(ASSOC_FILE, sep=\"\\s+\", usecols=[\"SNP\", \"F_A\", \"F_U\"])\n",
    "\n",
    "df = pd.merge(log_hybrid, assoc_adjusted, on=\"SNP\", how=\"right\")\n",
    "\n",
    "#merge freq with df\n",
    "freq_assoc = pd.merge(assoc, df, on=\"SNP\", how=\"left\")\n",
    "\n",
    "#read in recode file\n",
    "recode = pd.read_csv(RECODE_FILE, sep=\"\\s+\")\n",
    "\n",
    "# Pre-filter the dataset\n",
    "cases_data = recode[recode[\"PHENOTYPE\"] == 2]\n",
    "controls_data = recode[recode[\"PHENOTYPE\"] == 1]\n",
    "# Make a list from the column names\n",
    "column_names = recode.columns.tolist()\n",
    "\n",
    "# Drop the first 6 columns to keep the variants \n",
    "variants = column_names[6:]\n",
    "results = []\n",
    "for variant in variants:\n",
    "    # For cases\n",
    "    hom_cases = cases_data[cases_data[variant] == 2].shape[0]\n",
    "    het_cases = cases_data[cases_data[variant] == 1].shape[0]\n",
    "    total_cases = cases_data.shape[0]\n",
    "    # For controls\n",
    "    hom_controls = controls_data[controls_data[variant] == 2].shape[0]\n",
    "    het_controls = controls_data[controls_data[variant] == 1].shape[0]\n",
    "    total_controls = controls_data.shape[0]\n",
    "    results.append({\n",
    "        \"Variant\": variant,\n",
    "        \"Hom Cases\": hom_cases,\n",
    "        \"Het Cases\": het_cases,\n",
    "        \"Total Cases\": total_cases,\n",
    "        \"Hom Controls\": hom_controls,\n",
    "        \"Het Controls\": het_controls,\n",
    "        \"Total Controls\": total_controls,\n",
    "    })\n",
    "\n",
    "# Return results\n",
    "df_results = pd.DataFrame(results)\n",
    "df_results[\"SNP\"] = df_results[\"Variant\"].apply(lambda x: x.rsplit(\"_\", 1)[0])\n",
    "df_results = df_results.drop(\"Variant\", axis=1)\n",
    "\n",
    "# Merge with assoc results \n",
    "full_results = pd.merge(freq_assoc, df_results, on=\"SNP\", how=\"left\")\n",
    "\n",
    "# get BONF threshold value and append to df\n",
    "full_results[\"Bonferroni Threshold\"] = 0.05/len(full_results[\"SNP\"])\n",
    "\n",
    "##  get r2 for imputation \n",
    "! grep -v \"##\" {WORK_DIR}/{CHROM}_{ancestry}.pvar > {WORK_DIR}/{CHROM}_{ancestry}_imputation_info.txt\n",
    "\n",
    "variant_file = pd.read_csv(f\"{WORK_DIR}/{CHROM}_{ancestry}_imputation_info.txt\", sep =\"\\t\")\n",
    "R2 = variant_file[\"INFO\"].str.extract(\"(R2=\\d+(\\.\\d+)?)\", expand=False)[0].str.strip()\n",
    "\n",
    "variant_file[\"R2\"] = R2.str.strip(\"R2=\")\n",
    "variant_file[\"SNP\"] = variant_file[\"ID\"]\n",
    "# make df with variant R2\n",
    "variant_df = variant_file[[\"SNP\", \"R2\"]].copy()\n",
    "\n",
    "# append R2 to the final results dataframe\n",
    "full_results_with_r = pd.merge(full_results, variant_df, on=\"SNP\", how=\"left\")\n",
    "full_results_with_r\n",
    "\n",
    "#subset to only columns to keep\n",
    "clean_full_results = full_results_with_r[[\"SNP\", \"A1\", \"A1_FREQ\", \"P\", \"OR\", \"LOG(OR)_SE\",\"L95\", \"U95\", \"BONF\", \n",
    "                                   \"F_A\", \"F_U\",\n",
    "                                   \"Hom Cases\", \"Het Cases\", \"Total Cases\", \n",
    "                                   \"Hom Controls\",\"Het Controls\", \"Total Controls\",\n",
    "                                  \"Bonferroni Threshold\", \"R2\"]].copy()\n",
    "clean_full_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "6032d783-fcf5-4d4b-8d54-76114c8bfdca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add ancestry \n",
    "clean_full_results.insert(0, \"Ancestry\", ancestry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c544e80-b23d-4540-a342-ef166c925342",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "print(clean_full_results.shape)\n",
    "print(f\"No. of SNPs: {clean_full_results.shape[0]}\")\n",
    "clean_full_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c10409f2-a75e-4c50-88ca-be82c9057743",
   "metadata": {},
   "source": [
    "### Keep variants only in cases with CADD > 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88db0549-4822-449c-abb1-2edb8c31523a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# get variants that are only in cases\n",
    "cases_only = clean_full_results[(clean_full_results[\"Hom Controls\"] + clean_full_results[\"Het Controls\"]) == 0].copy()\n",
    "cases_only\n",
    "\n",
    "annotations = pd.read_csv(f\"{WORK_DIR}/{ancestry}_{GENE}_coding_nonsyn_variant_annotations.txt\", sep=\"\\t\", usecols=[\"SNP\", \"CADD_phred\", \"ExonicFunc.refGene\"])\n",
    "\n",
    "cases_only = cases_only.merge(annotations, on=\"SNP\")\n",
    "cases_only\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "154b499c-d40b-4738-a05e-df1f9c4641c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# keep case only variants with CADD score > 20\n",
    "cases_only_cadd20 = cases_only[cases_only[\"CADD_phred\"] > 20]\n",
    "only_in_cases_df = cases_only_cadd20[[\"Ancestry\", \"SNP\", \"ExonicFunc.refGene\", \"Hom Cases\", \"Het Cases\", \"Total Cases\", \"CADD_phred\", \"F_A\"]].copy()\n",
    "only_in_cases_df.rename(columns={\"F_A\":\"Case MAF\", \"ExonicFunc.refGene\":\"Consequence\"}, inplace=True)\n",
    "\n",
    "# write to file\n",
    "only_in_cases_df.to_csv(f\"{WORK_DIR}/{ancestry}_{GENE}_possibly_pathogenic_variants.txt\", index=False, header=True, sep=\"\\t\")\n",
    "only_in_cases_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7b24cc5-7b87-4aff-8fa9-6cbb473548ce",
   "metadata": {},
   "source": [
    "## Finalize Results Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20f59100-5cb5-4242-92b1-efac6245c749",
   "metadata": {},
   "outputs": [],
   "source": [
    "# append variant consequences\n",
    "clean_full_results_annotated = clean_full_results.merge(annotations, on=\"SNP\", how=\"left\")\n",
    "clean_full_results_annotated.rename(columns={\"ExonicFunc.refGene\":\"Consequence\"}, inplace=True)\n",
    "col = clean_full_results_annotated.pop(\"Consequence\")\n",
    "clean_full_results_annotated.insert(1, \"Consequence\", col)\n",
    "clean_full_results_annotated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ac311f5-aa43-4684-b715-f8fb09251f08",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Look at significant SNPs, if any\n",
    "sig_freq = clean_full_results_annotated[clean_full_results_annotated[\"P\"]<0.05]\n",
    "sig_snps = sig_freq[\"SNP\"].tolist()\n",
    "sig_df_results = clean_full_results_annotated[clean_full_results_annotated[\"SNP\"].isin(sig_snps)]\n",
    "sig_df_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "c08ef937-e929-4774-836d-77ced1835191",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save files to working directory\n",
    "clean_full_results_annotated.to_csv(f\"{WORK_DIR}/{ancestry}_{GENE}_GP2_R9.fullVariantInformation.txt\", sep=\"\\t\", index=False)\n",
    "sig_df_results.to_csv(f\"{WORK_DIR}/{ancestry}_{GENE}_GP2_R9.SignificantVariantInformation.txt\" , sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "python3",
   "name": "r-cpu.m125",
   "type": "gcloud",
   "uri": "us-docker.pkg.dev/deeplearning-platform-release/gcr.io/r-cpu:m125"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc-showcode": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
