{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1b71b59a",
   "metadata": {},
   "source": [
    "# AMP PD\n",
    "* **Project:** ADRD-SORL1-Biobanks\n",
    "* **Version:** Python/3.10\n",
    "* **Last Updated:** 14-Jun-2025\n",
    "\n",
    "## Notebook Overview\n",
    "Characterization of SORL1 variants, allele freqs, association analysis, burden analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62a109c1",
   "metadata": {},
   "source": [
    "# Query AMP PD to check for variants of interest, and allele frequency"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff1e223d",
   "metadata": {},
   "source": [
    "## Variables used \n",
    "- `${ANCESTRY}` = EUR, AFR, AMR, AAC, AJ, MDE, SAS, CAS, EAS, FIN, CAH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84960676",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bf7c51a7-e03e-4081-9f35-765991df5f0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>SEX</th>\n",
       "      <th>AGE_BASELINE</th>\n",
       "      <th>AGE_DIAGNOSIS</th>\n",
       "      <th>FAMILY_HISTORY</th>\n",
       "      <th>EDUCATION</th>\n",
       "      <th>LATEST_DX</th>\n",
       "      <th>DX_TYPE</th>\n",
       "      <th>ENROLL_STUDY_ARM</th>\n",
       "      <th>COHORT</th>\n",
       "      <th>PD_PHENO</th>\n",
       "      <th>PD_EXTRA_PHENO</th>\n",
       "      <th>DLB_PHENO</th>\n",
       "      <th>MSA_PHENO</th>\n",
       "      <th>PSP_PHENO</th>\n",
       "      <th>CBD_PHENO</th>\n",
       "      <th>AGE_ANALYSIS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BF-1001</td>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No PD Nor Other Neurological Disorder</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BF-1002</td>\n",
       "      <td>2</td>\n",
       "      <td>66</td>\n",
       "      <td>61.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BF-1003</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>56.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BF-1004</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BF-1005</td>\n",
       "      <td>2</td>\n",
       "      <td>61</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No PD Nor Other Neurological Disorder</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ID  SEX  AGE_BASELINE  AGE_DIAGNOSIS  FAMILY_HISTORY  EDUCATION  \\\n",
       "0  BF-1001    1            55            NaN             0.0        1.0   \n",
       "1  BF-1002    2            66           61.0             0.0        1.0   \n",
       "2  BF-1003    1            61           56.0             0.0        1.0   \n",
       "3  BF-1004    1            62           55.0             0.0        1.0   \n",
       "4  BF-1005    2            61            NaN             0.0        1.0   \n",
       "\n",
       "                               LATEST_DX  DX_TYPE ENROLL_STUDY_ARM   COHORT  \\\n",
       "0  No PD Nor Other Neurological Disorder      NaN  Healthy Control  BIOFIND   \n",
       "1                          Idiopathic PD      NaN               PD  BIOFIND   \n",
       "2                          Idiopathic PD      NaN               PD  BIOFIND   \n",
       "3                          Idiopathic PD      NaN               PD  BIOFIND   \n",
       "4  No PD Nor Other Neurological Disorder      NaN  Healthy Control  BIOFIND   \n",
       "\n",
       "   PD_PHENO  PD_EXTRA_PHENO  DLB_PHENO  MSA_PHENO  PSP_PHENO  CBD_PHENO  \\\n",
       "0         1               1          1          1          1          1   \n",
       "1         2               2         -9         -9         -9         -9   \n",
       "2         2               2         -9         -9         -9         -9   \n",
       "3         2               2         -9         -9         -9         -9   \n",
       "4         1               1          1          1          1          1   \n",
       "\n",
       "   AGE_ANALYSIS  \n",
       "0          55.0  \n",
       "1          61.0  \n",
       "2          56.0  \n",
       "3          55.0  \n",
       "4          61.0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qc_covar_PD = pd.read_csv(\"/${WORK_DIR}/addedPHENO_COVFILE_releasev3_SEPT2022_fromTerra.csv\", sep=\",\")\n",
    "qc_covar_PD.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "633ec20d-b784-48d5-a164-d3f451b7df4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ancestries = [\"AFR\", \"AJ\", \"EUR\", \"MDE\", \"AAC\", \"AMR\", \"FIN\", \"CAS\", \"SAS\", \"EAS\", \"CAH\"]\n",
    "\n",
    "qc_covar_PD = pd.read_csv(\"/${WORK_DIR}/addedPHENO_COVFILE_releasev3_SEPT2022_fromTerra.csv\", sep=\",\")\n",
    "\n",
    "all_to_remove = set()\n",
    "\n",
    "for ancestry in ancestries:\n",
    "    remove_file = f\"/${WORK_DIR}/toRemove_1stand2ndDegree_Relateds_{ancestry}_noDups.txt\"\n",
    "    \n",
    "    try:\n",
    "        relateds = pd.read_csv(remove_file, sep=\"\\t\", header=None, names=[\"FID\", \"IID\"])\n",
    "        all_to_remove.update(relateds[\"IID\"].tolist())\n",
    "        print(f\"Loaded {len(relateds)} samples to remove from {ancestry}\")\n",
    "    except FileNotFoundError:\n",
    "        print(f\"‚ö†Ô∏è File not found for ancestry: {ancestry} ‚Äî skipping.\")\n",
    "\n",
    "filtered_covar = qc_covar_PD[~qc_covar_PD['ID'].isin(all_to_remove)]\n",
    "\n",
    "filtered_covar.to_csv(\"qc_covar_PD_unrelateds_only.csv\", sep=\",\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "162ffdb5-7d35-4c11-a7af-8b5daac01d2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>SEX</th>\n",
       "      <th>AGE_BASELINE</th>\n",
       "      <th>AGE_DIAGNOSIS</th>\n",
       "      <th>FAMILY_HISTORY</th>\n",
       "      <th>EDUCATION</th>\n",
       "      <th>LATEST_DX</th>\n",
       "      <th>DX_TYPE</th>\n",
       "      <th>ENROLL_STUDY_ARM</th>\n",
       "      <th>COHORT</th>\n",
       "      <th>PD_PHENO</th>\n",
       "      <th>PD_EXTRA_PHENO</th>\n",
       "      <th>DLB_PHENO</th>\n",
       "      <th>MSA_PHENO</th>\n",
       "      <th>PSP_PHENO</th>\n",
       "      <th>CBD_PHENO</th>\n",
       "      <th>AGE_ANALYSIS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BF-1001</td>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No PD Nor Other Neurological Disorder</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BF-1002</td>\n",
       "      <td>2</td>\n",
       "      <td>66</td>\n",
       "      <td>61.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BF-1003</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>56.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BF-1004</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BF-1005</td>\n",
       "      <td>2</td>\n",
       "      <td>61</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No PD Nor Other Neurological Disorder</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ID  SEX  AGE_BASELINE  AGE_DIAGNOSIS  FAMILY_HISTORY  EDUCATION  \\\n",
       "0  BF-1001    1            55            NaN             0.0        1.0   \n",
       "1  BF-1002    2            66           61.0             0.0        1.0   \n",
       "2  BF-1003    1            61           56.0             0.0        1.0   \n",
       "3  BF-1004    1            62           55.0             0.0        1.0   \n",
       "4  BF-1005    2            61            NaN             0.0        1.0   \n",
       "\n",
       "                               LATEST_DX  DX_TYPE ENROLL_STUDY_ARM   COHORT  \\\n",
       "0  No PD Nor Other Neurological Disorder      NaN  Healthy Control  BIOFIND   \n",
       "1                          Idiopathic PD      NaN               PD  BIOFIND   \n",
       "2                          Idiopathic PD      NaN               PD  BIOFIND   \n",
       "3                          Idiopathic PD      NaN               PD  BIOFIND   \n",
       "4  No PD Nor Other Neurological Disorder      NaN  Healthy Control  BIOFIND   \n",
       "\n",
       "   PD_PHENO  PD_EXTRA_PHENO  DLB_PHENO  MSA_PHENO  PSP_PHENO  CBD_PHENO  \\\n",
       "0         1               1          1          1          1          1   \n",
       "1         2               2         -9         -9         -9         -9   \n",
       "2         2               2         -9         -9         -9         -9   \n",
       "3         2               2         -9         -9         -9         -9   \n",
       "4         1               1          1          1          1          1   \n",
       "\n",
       "   AGE_ANALYSIS  \n",
       "0          55.0  \n",
       "1          61.0  \n",
       "2          56.0  \n",
       "3          55.0  \n",
       "4          61.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qc_covar_PD2 = pd.read_csv(\"qc_covar_PD_unrelateds_only.csv\", sep=\",\")\n",
    "qc_covar_PD2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61a6b1a8-b1e7-4f9e-882d-9b3a7d24c57b",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_case_PD2 = qc_covar_PD2[qc_covar_PD2[\"PD_PHENO\"]==2]\n",
    "qc_case_PD2.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7259f024-82c0-4756-a85a-a8585dd3888d",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_case_PD2_plink = qc_case_PD2[[\"ID\"]]\n",
    "qc_case_PD2_plink.to_csv(\"qc_case_PD2_plink.txt\", sep=\",\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fd94cbb8-a12d-4a70-b215-e812583faf9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "!awk '{print $0, $1}' qc_case_PD2_plink.txt > qc_case_PD2_ID_ID_plink.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74a7c3ea-0fd2-416b-882a-8559de12e7a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_control_PD2 = qc_covar_PD2[qc_covar_PD2[\"PD_PHENO\"]==1]\n",
    "qc_control_PD2.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0c281377-cd60-4aad-88d1-2a03771def44",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_control_PD2_plink = qc_control_PD2[[\"ID\"]]\n",
    "qc_control_PD2_plink.to_csv(\"qc_control_PD2_plink.txt\", sep=\",\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bcc778fb-0280-4987-97cf-58df47c12234",
   "metadata": {},
   "outputs": [],
   "source": [
    "!awk '{print $0, $1}' qc_control_PD2_plink.txt > qc_control_PD2_ID_ID_plink.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c691a10-f50c-4078-86c4-efb9b30d74e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink\n",
    "plink2 --pfile /${WORK_DIR}/FILTERED.AMP_PD_${ANCESTRY} \\\n",
    "--chr 11 --from-bp 121452314 --to-bp 121633763 --make-bed --out AMPPD_Alex_vars_all_${ANCESTRY}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e8ece3c-2e5d-42b8-b942-399e82176219",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink/1.9\n",
    "\n",
    "plink --bfile AMPPD_Alex_vars_all_${ANCESTRY} --keep qc_case_PD2_ID_ID_plink.txt --make-bed --out AMPPD_Alex_vars_all_${ANCESTRY}_cases2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ac8d7be-1ea9-424e-8e35-ffadfb652dbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink/1.9\n",
    "\n",
    "plink --bfile AMPPD_Alex_vars_all_${ANCESTRY}_cases2 --freq --out AMPPD_Alex_vars_all_${ANCESTRY}_cases2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9310d677-36e8-4526-8816-45e0c6167e4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink/1.9\n",
    "\n",
    "plink --bfile AMPPD_Alex_vars_all_${ANCESTRY} --keep qc_control_PD2_ID_ID_plink.txt --make-bed --out AMPPD_Alex_vars_all_${ANCESTRY}_controls2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa712e65-893e-482f-bc3b-f1a93e02ba71",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink/1.9\n",
    "\n",
    "plink --bfile AMPPD_Alex_vars_all_${ANCESTRY}_controls2 --freq --out AMPPD_Alex_vars_all_${ANCESTRY}_controls2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f56f474-3fdd-4eb6-8e2e-e8c44409fecb",
   "metadata": {},
   "source": [
    "## Check Zygosity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51c90742-5c2b-4697-b8e8-62475316dd1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink\n",
    "plink2 --bfile AMPPD_Alex_vars_all_${ANCESTRY} --recode A --out AMPPD_Alex_vars_all_${ANCESTRY}_recode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78489157-98cb-4d88-878f-62a27e70e685",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink\n",
    "plink2 --bfile AMPPD_Alex_vars_all_${ANCESTRY}_cases2 --recode A --out AMPPD_Alex_vars_all_${ANCESTRY}_cases2_recode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd5d1943-f741-4ac7-954b-4c8c1f460da9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink\n",
    "plink2 --bfile AMPPD_Alex_vars_all_${ANCESTRY}_controls2 --recode A --out AMPPD_Alex_vars_all_${ANCESTRY}_controls2_recode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daf44629-8f96-48b6-bc2f-e107bdd22665",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Merge files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d356345-7420-4db0-a9b4-53a57a43de6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "control_amp_paths = glob.glob(\"AMPPD_*controls2*.frq\")\n",
    "case_amp_paths = glob.glob(\"AMPPD_*cases2*.frq\")\n",
    "\n",
    "control_amp_paths.sort()\n",
    "case_amp_paths.sort()\n",
    "\n",
    "amp_paths = list(zip(control_amp_paths, case_amp_paths))\n",
    "print(f\"Total paired files: {len(amp_paths)}\")\n",
    "\n",
    "df_list = []\n",
    "for paths in amp_paths:\n",
    "    for i, path in enumerate(paths):\n",
    "        df = pd.read_csv(path, sep=r'\\s+', engine='python')\n",
    "\n",
    "        df[\"NCHROBS\"] = pd.to_numeric(df[\"NCHROBS\"])\n",
    "        nchrobs = df[\"NCHROBS\"].max()\n",
    "        ancestry = path.split(\"_\")[4]\n",
    "        maf_col_name = f\"{ancestry} {'Controls' if i == 0 else 'Cases'} MAF (NCHROBS = {nchrobs})\"\n",
    "        df.rename({\"MAF\": maf_col_name}, inplace=True, axis=1)\n",
    "\n",
    "        df = df[[\"SNP\", maf_col_name]]\n",
    "        display(df.head(3))\n",
    "\n",
    "        df_list.append(df)\n",
    "\n",
    "for i in range(len(df_list)):\n",
    "    print(f\"Match with first file [{i}]:\", list(df_list[0][\"SNP\"]) == list(df_list[i][\"SNP\"]))\n",
    "\n",
    "merged_df = df_list[0]\n",
    "for df in df_list[1:]:\n",
    "    merged_df = pd.merge(merged_df, df, on='SNP', how=\"outer\")\n",
    "\n",
    "merged_df = merged_df.fillna(0)\n",
    "\n",
    "display(merged_df.head())\n",
    "\n",
    "os.makedirs(\"Merged_AMPPD\", exist_ok=True)\n",
    "merged_df.to_csv(\"Merged_AMPPD/AMPPD2.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ad93537",
   "metadata": {},
   "source": [
    "## Count the number of cases and controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "5567d4f5-e668-42e6-9791-136fcffc30bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "! awk 'NR==FNR {ids[$1]; next} $1 in ids' qc_case_PD2_ID_ID_plink.txt /${WORK_DIR}/FILTERED.AMP_PD_ancestry_${ANCESTRY}.samples > filtered_samples2_${ANCESTRY}.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70c60f8c-5f55-4316-acb2-c53ab3b900f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc filtered_samples2_${ANCESTRY}.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "deb1088b-7787-4f69-9730-e0c209c1a6de",
   "metadata": {},
   "outputs": [],
   "source": [
    "! awk 'NR==FNR {ids[$1]; next} $1 in ids' qc_control_PD2_ID_ID_plink.txt /${WORK_DIR}/FILTERED.AMP_PD_ancestry_${ANCESTRY}.samples > filtered_samples_controls2_${ANCESTRY}.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17102e48-97ed-4ed4-91c0-2a5b6615414b",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc filtered_samples_controls2_${ANCESTRY}.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7331d99a-285f-4164-8d39-f125b77ad1c4",
   "metadata": {},
   "source": [
    "## Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5fe5eb0-aa60-465a-a088-56abb539fcf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load annovar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79bf41da",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "table_annovar.pl AMP_DLB_8.cleaned.vcf $ANNOVAR_DATA/hg38 \\\n",
    "    --buildver hg38 \\\n",
    "    --remove \\\n",
    "    --thread 48 \\\n",
    "    --maxgenethread 48 \\\n",
    "    --protocol refGene,clinvar_20140902,avsnp151,dbnsfp47a \\\n",
    "    --operation g,f,f,f \\\n",
    "    --nopolish \\\n",
    "    --nastring . \\\n",
    "    --out AMP_DLB_8.vcf.anno \\\n",
    "    --vcfinput"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09392878",
   "metadata": {},
   "source": [
    "## Merge files and apply criteria (exonic and splicing variants, mac 2, CADD>20, and Only present in cases)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "3bc6aa7a-d676-451e-be12-b68fbc8e0f50",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"AMP_PD_8.vcf.anno.hg38_multianno.txt\", sep=\"\\t\")\n",
    "\n",
    "columns_to_keep = [\n",
    "    \"Chr\", \"Start\", \"End\", \"Ref\", \"Alt\",\n",
    "    \"Func.refGene\", \"Gene.refGene\", \"GeneDetail.refGene\",\n",
    "    \"ExonicFunc.refGene\", \"AAChange.refGene\",\n",
    "    \"clinvar_20140902\", \"avsnp151\", \"CADD_phred\"\n",
    "]\n",
    "\n",
    "filtered_df = df[columns_to_keep]\n",
    "\n",
    "filtered_df.to_csv(\"AMP_PD_8.filtered.txt\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "7f0f4516-cd9d-42e5-8444-1b4e0d9a95c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "anno = pd.read_csv('AMP_PD_8.filtered.txt', sep='\\t')\n",
    "freq = pd.read_csv('Merged_AMPPD/AMPPD2.csv')\n",
    "\n",
    "freq['Start'] = freq['SNP'].apply(lambda x: int(x.split(':')[1]))\n",
    "\n",
    "merged = pd.merge(anno, freq, on='Start', how='inner')\n",
    "\n",
    "merged.to_csv('AMP_PD_8_merged_output2.txt', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe056aab-47df-495a-9269-1056e3f15bb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "ancestries = [\"AFR\",\"AJ\",\"EUR\",\"MDE\",\"AAC\",\"AMR\",\"FIN\",\"CAS\",\"SAS\",\"EAS\",\"CAH\"]\n",
    "zyg_list = []\n",
    "\n",
    "for ancestry in ancestries:\n",
    "    for group in [\"cases\", \"controls\"]:\n",
    "        filename = f\"AMPPD_Alex_vars_all_{ancestry}_{group}2_recode.raw\"  \n",
    "        if os.path.exists(filename):\n",
    "            print(f\"Processing: {filename}\")\n",
    "            df = pd.read_csv(filename, sep=\"\\t\")\n",
    "            df.drop(columns=[\"FID\", \"IID\", \"PAT\", \"MAT\", \"SEX\", \"PHENOTYPE\"], inplace=True)\n",
    "            df = df.fillna(2).astype(int)\n",
    "\n",
    "            snps = [snpid.split(\"_\")[0] for snpid in df.columns]\n",
    "            hom_alt = (df == 0).sum()\n",
    "            het = (df == 1).sum()\n",
    "\n",
    "            df_zyg = pd.DataFrame({\n",
    "                \"SNP\": snps,\n",
    "                f\"{ancestry}_{group}_hom_alt_ac\": hom_alt.values,\n",
    "                f\"{ancestry}_{group}_het_ac\": het.values\n",
    "            })\n",
    "\n",
    "            zyg_list.append(df_zyg)\n",
    "        else:\n",
    "            print(f\"Missing file: {filename}\")\n",
    "\n",
    "\n",
    "from functools import reduce\n",
    "\n",
    "if zyg_list:\n",
    "    merged_zyg = reduce(lambda left, right: pd.merge(left, right, on='SNP', how='outer'), zyg_list)\n",
    "\n",
    "    \n",
    "    merged = pd.read_csv(\"AMP_PD_8_merged_output2.txt\", sep=\"\\t\")  \n",
    "    final_merged = pd.merge(merged, merged_zyg, on='SNP', how='left')\n",
    "    final_merged.to_csv(\"AMP_PD_8_final_output2.txt\", sep=\"\\t\", index=False)\n",
    "    print(\"Final merged file written: AMP_PD_8_final_output2.txt\")\n",
    "else:\n",
    "    print(\"No valid zygosity data found to merge.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b1fa72d-3551-45c8-a447-0d4714439554",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_amp_merged = pd.read_csv(\"AMP_PD_8_final_output2.txt\", sep=\"\\t\")\n",
    "\n",
    "amp_freq_cols = df_amp_merged.filter(like=\"NCHROBS\").columns\n",
    "\n",
    "criteria_amp_anno = (\n",
    "    (df_amp_merged[\"Func.refGene\"].isin([\"splicing\", \"exonic\"])) &\n",
    "    ~(df_amp_merged[\"ExonicFunc.refGene\"] == \"synonymous SNV\")\n",
    ")\n",
    "\n",
    "df_amp_filtered = df_amp_merged[criteria_amp_anno]\n",
    "\n",
    "criteria_amp_freq = (df_amp_filtered[amp_freq_cols] > 0.0).any(axis=1)\n",
    "df_amp_filtered = df_amp_filtered[criteria_amp_freq]\n",
    "\n",
    "df_amp_filtered.to_csv(\"AMP_PD_8_filtered_coding_output2.txt\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d17181df-98ed-4301-b24b-656343fb4d8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc AMP_PD_8_filtered_coding_output2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a173043d-8ba0-4c1c-8de4-d92333230b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_amp_filtered = pd.read_csv(\"AMP_PD_8_filtered_coding_output2.txt\", sep=\"\\t\")\n",
    "\n",
    "amp_controls_cols = df_amp_filtered.filter(like=\"Controls MAF\").columns\n",
    "\n",
    "df_amp_filtered_casesonly = df_amp_filtered[(df_amp_filtered[amp_controls_cols] == 0.0).all(axis=1)]\n",
    "\n",
    "df_amp_filtered_casesonly.to_csv(\"AMP_PD_8__coding_casesonly_output2.txt\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d870804-6184-436e-a107-757184de1dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc AMP_PD_8__coding_casesonly_output2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd307e91-2a64-4254-974f-f75a9b6a7550",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_amp_casesonly = pd.read_csv(\"AMP_PD_8__coding_casesonly_output2.txt\", sep=\"\\t\")\n",
    "\n",
    "criteria_amp_cadd = (df_amp_casesonly[\"CADD_phred\"] >= 20) | (df_amp_casesonly[\"CADD_phred\"].isna())\n",
    "df_amp_casesonly_cadd_filtered = df_amp_casesonly[criteria_amp_cadd]\n",
    "\n",
    "df_amp_casesonly_cadd_filtered.to_csv(\"AMP_PD_8__coding_casesonly_CADD20_output2.txt\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73be4a77-97cc-4675-8f74-924b2a392d23",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc AMP_PD_8__coding_casesonly_CADD20_output2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7ef94d5-de9b-4963-98b9-72adeefc28db",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_amp_casesonly_cadd = pd.read_csv(\"AMP_PD_8__coding_casesonly_CADD20_output2.txt\", sep=\"\\t\")\n",
    "\n",
    "criteria_${ANCESTRY}_ac = (df_amp_casesonly_cadd[\"${ANCESTRY}_cases_het_ac\"] >= 2) | (df_amp_casesonly_cadd[\"${ANCESTRY}_cases_hom_alt_ac\"] >= 2)\n",
    "df_amp_casesonly_cadd_${ANCESTRY}_filtered = df_amp_casesonly_cadd[criteria_${ANCESTRY}_ac]\n",
    "\n",
    "df_amp_casesonly_cadd_${ANCESTRY}_filtered.to_csv(\"AMP_PD_8__coding_casesonly_CADD20_${ANCESTRY}AC2_output2.txt\", sep=\"\\t\", index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d109f7d9-9cf9-40c7-bd3c-ade2f2620893",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc AMP_PD_8__coding_casesonly_CADD20_${ANCESTRY}AC2_output2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb5e8c88-124b-472b-9173-d6342be5a0a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "! cat AMP_PD_8__coding_casesonly_CADD20_${ANCESTRY}AC2_output2.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20b02547-5d54-4269-8c2a-55bcc10662b7",
   "metadata": {},
   "source": [
    "## Caunt total variansts per ancestry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5d02cc0-88f9-40b7-adc0-5ffd0a1590a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"AMP_PD_8_merged_output2.txt\", sep=\"\\t\")\n",
    "\n",
    "ancestries = [\"AAC\", \"AFR\", \"AJ\", \"AMR\", \"CAH\", \"CAS\", \"EAS\", \"EUR\", \"FIN\", \"MDE\", \"SAS\"]\n",
    "\n",
    "summary_tables = []\n",
    "\n",
    "for ancestry in ancestries:\n",
    "    control_col = [col for col in df.columns if col.startswith(f\"{ancestry} Controls MAF\")][0]\n",
    "    case_col = [col for col in df.columns if col.startswith(f\"{ancestry} Cases MAF\")][0]\n",
    "\n",
    "    ancestry_variants = df[(df[control_col] > 0) | (df[case_col] > 0)].copy()\n",
    "\n",
    "    if not ancestry_variants.empty:\n",
    "        func_counts = ancestry_variants.groupby(\"Func.refGene\").size().reset_index(name=\"VariantCount\")\n",
    "        func_counts[\"Ancestry\"] = ancestry\n",
    "\n",
    "        exonic_variants = ancestry_variants[ancestry_variants[\"Func.refGene\"] == \"exonic\"]\n",
    "        exonic_counts = exonic_variants.groupby(\"ExonicFunc.refGene\").size().reset_index(name=\"VariantCount\")\n",
    "        exonic_counts[\"Ancestry\"] = ancestry\n",
    "        exonic_counts[\"Func.refGene\"] = \"exonic\"\n",
    "\n",
    "        summary_tables.append(func_counts)\n",
    "        summary_tables.append(exonic_counts)\n",
    "\n",
    "final_summary = pd.concat(summary_tables, ignore_index=True)\n",
    "\n",
    "final_summary.to_csv(\"AMP_PD_variant_ancestry_func_and_exonic_type_counts2.txt\", sep=\"\\t\", index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88fd10ac-b073-40b6-915e-5d712727064d",
   "metadata": {},
   "outputs": [],
   "source": [
    "! head AMP_PD_variant_ancestry_func_and_exonic_type_counts2.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4adf9ae-8b87-40f2-8ab6-7def5e580677",
   "metadata": {},
   "source": [
    "## Association analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2af2c938-3a0b-439c-ae4b-d8c0738ce171",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "from functools import reduce\n",
    "import argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cac681b2-1ff2-464a-8d9c-967895e4ab78",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink\n",
    "plink2 --bfile AMPPD_Alex_vars_all_${ANCESTRY} --mac 2 --make-bed --out  AMPPD_Alex_vars_all_MAC2_${ANCESTRY}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca9f1173-5042-4d95-b4f4-e59c95bbd322",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink\n",
    "plink2 --bfile AMPPD_Alex_vars_all_MAC2_${ANCESTRY} --recode vcf-iid --out AMPPD_Alex_vars_all_MAC2_${ANCESTRY}_recode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "45618d1e-25e0-4194-b9aa-687a09164134",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gzip AMPPD_Alex_vars_all_MAC2_${ANCESTRY}_recode.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "65fe37fe-559c-4d5b-9984-20ca52245477",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"AMP_PD_8.vcf.anno.hg38_multianno.txt\", sep='\\t')\n",
    "\n",
    "filtered_df = df[df['Func.refGene'].isin(['splicing', 'exonic'])]\n",
    "\n",
    "filtered_df.to_csv(\"AMP_PD_8_splicing_exonic.txt\", sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a1e8352-5ef6-411d-8344-acdff91759b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc AMP_PD_8_splicing_exonic.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f38fd8db-9818-41f1-8741-98a3b8fcbb20",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "filter_df = pd.read_csv(\"AMP_PD_8_splicing_exonic.txt\", sep='\\t', usecols=[\"Chr\", \"Start\", \"Ref\", \"Alt\"])\n",
    "\n",
    "filter_df['Chr'] = filter_df['Chr'].astype(str).str.replace('chr', '', regex=False)\n",
    "\n",
    "filter_set = set(zip(filter_df['Chr'], filter_df['Start'].astype(str), filter_df['Ref'], filter_df['Alt']))\n",
    "\n",
    "ancestries = [\"EUR\", \"AFR\", \"AMR\", \"EAS\", \"SAS\", \"AAC\", \"AJ\", \"CAS\", \"CAH\", \"FIN\", \"MDE\"]\n",
    "\n",
    "for ancestry in ancestries:\n",
    "    input_vcf = f\"AMPPD_Alex_vars_all_MAC2_{ancestry}_recode.vcf\" \n",
    "    output_vcf = f\"AMPPD_Alex_vars_splicing_exonic_{ancestry}.vcf\"\n",
    "\n",
    "    with open(input_vcf, 'r') as infile, open(output_vcf, 'w') as outfile:\n",
    "        for line in infile:\n",
    "            if line.startswith(\"#\"):\n",
    "                outfile.write(line)  \n",
    "            else:\n",
    "                cols = line.strip().split(\"\\t\")\n",
    "                chrom, pos, ref, alt = cols[0], cols[1], cols[3], cols[4]\n",
    "                if (chrom, pos, ref, alt) in filter_set:\n",
    "                    outfile.write(line)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a9eabf4-063b-43ee-a7c7-49b3ba3fefd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc AMPPD_Alex_vars_splicing_exonic_${ANCESTRY}.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "466ebd0d-effe-4b59-b772-5c4a6ecacb74",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gzip AMPPD_Alex_vars_splicing_exonic_${ANCESTRY}.vcf "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c96edba-10d9-4c86-a341-081a76b16ac7",
   "metadata": {},
   "source": [
    "## Creating pheno file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c99f629c-062e-4bf2-ba8d-3cb86e144410",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "ancestries = [\"AFR\", \"AJ\", \"EUR\", \"MDE\", \"AAC\", \"AMR\", \"FIN\", \"CAS\", \"SAS\", \"EAS\", \"CAH\"]\n",
    "\n",
    "covariate_file = \"/${WORK_DIR}/fromTerra/addedPHENO_COVFILE_releasev3_SEPT2022_fromTerra.csv\"\n",
    "cov_df = pd.read_csv(covariate_file)\n",
    "\n",
    "sample_id_column = 'ID'\n",
    "pheno_column = 'PD_PHENO'\n",
    "\n",
    "for ancestry in ancestries:\n",
    "    sample_file = f\"/${WORK_DIR}/FILTERED.AMP_PD_ancestry_{ancestry}.samples\"\n",
    "    remove_file = f\"/${WORK_DIR}/toRemove_1stand2ndDegree_Relateds_{ancestry}_noDups.txt\"\n",
    "    output_file = f\"FILTERED.AMP_PD_ancestry_{ancestry}_PHENO.txt\"\n",
    "\n",
    "    if not os.path.isfile(sample_file):\n",
    "        print(f\"‚ö†Ô∏è Sample file not found: {sample_file}\")\n",
    "        continue\n",
    "\n",
    "    samples = pd.read_csv(sample_file, sep=\"\\t\", header=None, names=[\"FID\", \"IID\"])\n",
    "\n",
    "    if os.path.isfile(remove_file):\n",
    "        to_remove = pd.read_csv(remove_file, sep=\"\\t\", header=None, names=[\"FID\", \"IID\"])\n",
    "        before = len(samples)\n",
    "        samples = samples[~samples[\"IID\"].isin(to_remove[\"IID\"])]\n",
    "        print(f\"üßπ {ancestry}: {before - len(samples)} related samples removed.\")\n",
    "    else:\n",
    "        print(f\"‚ö†Ô∏è No related file found for {ancestry}, skipping related filtering.\")\n",
    "\n",
    "    merged = samples.merge(cov_df[[sample_id_column, pheno_column]],\n",
    "                           left_on=\"IID\", right_on=sample_id_column, how=\"left\")\n",
    "\n",
    "    filtered = merged[merged[pheno_column].isin([1, 2])]\n",
    "\n",
    "    final = filtered[[\"FID\", \"IID\", pheno_column]]\n",
    "    final.columns = [\"FID\", \"IID\", \"PHENO\"]\n",
    "\n",
    "    final.to_csv(output_file, sep=\"\\t\", index=False)\n",
    "\n",
    "    print(f\"‚úÖ Ancestry {ancestry}: {len(final)} samples saved to {output_file}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4838dcb2-1a72-492c-be51-38dd514fd1fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "covar_file = \"qc_covar_PD_unrelateds_only.csv\"\n",
    "covar_df = pd.read_csv(covar_file)\n",
    "\n",
    "pca_files = {\n",
    "    \"AJ\": \"/${WORK_DIR}/PCA.FILTERED.AMP_PD_AJ.PD.eigenvec\",\n",
    "    \"EUR\": \"/${WORK_DIR}/PCA.FILTERED.AMP_PD_EUR.PD.eigenvec\"\n",
    "}\n",
    "\n",
    "merged_list = []\n",
    "\n",
    "for ancestry, pca_path in pca_files.items():\n",
    "    pca_df = pd.read_csv(pca_path, delim_whitespace=True)\n",
    "\n",
    "    if '#FID' in pca_df.columns:\n",
    "        pca_df.rename(columns={'#FID': 'FID'}, inplace=True)\n",
    "\n",
    "    merged = covar_df.merge(pca_df, left_on=\"ID\", right_on=\"IID\", how=\"inner\")\n",
    "\n",
    "    merged[\"Ancestry\"] = ancestry\n",
    "\n",
    "    merged_list.append(merged)\n",
    "\n",
    "    print(f\"‚úÖ {ancestry}: Matched {len(merged)} samples.\")\n",
    "\n",
    "final_merged = pd.concat(merged_list, ignore_index=True)\n",
    "\n",
    "final_merged['FID'] = final_merged['ID']\n",
    "final_merged['IID'] = final_merged['ID']\n",
    "\n",
    "cols = ['FID', 'IID'] + [col for col in final_merged.columns if col not in ['FID', 'IID']]\n",
    "final_merged = final_merged[cols]\n",
    "\n",
    "final_merged.to_csv(\"qc_covar_PD_unrelateds_with_PCs.txt\", sep='\\t', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67ce3b6e-309a-4e1c-9059-a70d61900f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"qc_covar_PD_unrelateds_with_PCs.csv\")\n",
    "\n",
    "missing_counts = df[['AGE_BASELINE', 'AGE_DIAGNOSIS']].isnull().sum()\n",
    "\n",
    "print(\"üîç Missing data summary:\")\n",
    "print(missing_counts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98496ab1-9839-48a6-a4bd-c2b75edd62ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"qc_covar_PD_unrelateds_with_PCs.csv\")\n",
    "\n",
    "df['AGE'] = df.apply(\n",
    "    lambda row: row['AGE_DIAGNOSIS'] if row['PD_PHENO'] == 2 else row['AGE_BASELINE'] if row['PD_PHENO'] == 1 else None,\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "df.to_csv(\"qc_covar_PD_unrelateds_with_PCs_with_AGE.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0653cac1-4b69-4b01-b4ab-03091cd2601d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"qc_covar_PD_unrelateds_with_PCs_with_AGE.csv\")\n",
    "\n",
    "missing_counts = df[['AGE']].isnull().sum()\n",
    "\n",
    "print(\"üîç Missing data summary:\")\n",
    "print(missing_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3b33482-ca43-4775-a591-ed705660c067",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "covar_file = \"qc_covar_PD_unrelateds_with_PCs.txt\"\n",
    "covar_df = pd.read_csv(covar_file, sep=\"\\t\")  \n",
    "\n",
    "covar_df.columns = covar_df.columns.str.strip()\n",
    "\n",
    "columns_to_keep = ['FID', 'IID', 'SEX', 'AGE_BASELINE', 'PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'PC6', 'PC7', 'PC8', 'PC9', 'PC10']\n",
    "\n",
    "filtered_df = covar_df[columns_to_keep]\n",
    "\n",
    "filtered_df.to_csv(\"qc_covar_PD_unrelateds_filtered.txt\", sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3205a735-3e11-42ee-8029-054d52b134df",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink/2.0\n",
    "\n",
    "vcf_file=\"/${WORK_DIR}/AMPPD_Alex_vars_splicing_exonic_EUR.vcf.gz\"\n",
    "\n",
    "ancestries=(\"EUR\")\n",
    "\n",
    "output_dir=\"/${WORK_DIR}\"\n",
    "\n",
    "mkdir -p \"$output_dir\"\n",
    "\n",
    "for ancestry in \"${ancestries[@]}\"; do\n",
    "  echo \"Processing ancestry: $ancestry\"\n",
    "\n",
    "  plink2 \\\n",
    "    --vcf \"$vcf_file\" \\\n",
    "    --double-id \\\n",
    "    --pheno \"FILTERED.AMP_PD_ancestry_${ancestry}_PHENO.txt\" \\\n",
    "    --adjust \\\n",
    "    --ci 0.95 \\\n",
    "    --covar \"qc_covar_PD_unrelateds_filtered.txt\" \\\n",
    "    --covar-name SEX,AGE_BASELINE,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10 \\\n",
    "    --threads 15 \\\n",
    "    --covar-variance-standardize \\\n",
    "    --out \"${output_dir}/Logistic_FID_IID_PHENO_case_controls_${ancestry}_AMPPD_chr11_Alex\" \\\n",
    "    --glm omit-ref firth-fallback cols=+a1freq,+a1freqcc,+a1count,+totallele,+a1countcc,+totallelecc,+gcountcc,+err \\\n",
    "    --silent\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "246f6780-e3ff-48a1-9519-e336711463d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "\n",
    "file_path = \"Logistic_FID_IID_PHENO_case_controls_EUR_AMPPD_chr11_Alex.PHENO.glm.logistic.hybrid\"\n",
    "df = pd.read_csv(file_path, sep=\"\\t\")\n",
    "\n",
    "filtered_df = df[df['TEST'] == 'ADD']\n",
    "\n",
    "filtered_df.to_csv(\"Logistic_FID_IID_PHENO_case_controls_EUR_AMPPD_chr11_Alex.PHENO.glm.logistic.hybrid.ADD_only.txt\", sep='\\t', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f42a2650-8067-426f-ad5e-5c9d09a9a236",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink/2.0\n",
    "\n",
    "vcf_file=\"/${WORK_DIR}/AMPPD_Alex_vars_splicing_exonic_AJ.vcf.gz\"\n",
    "\n",
    "ancestries=(\"AJ\")\n",
    "\n",
    "output_dir=\"/${WORK_DIR}\"\n",
    "\n",
    "mkdir -p \"$output_dir\"\n",
    "\n",
    "for ancestry in \"${ancestries[@]}\"; do\n",
    "  echo \"Processing ancestry: $ancestry\"\n",
    "\n",
    "  plink2 \\\n",
    "    --vcf \"$vcf_file\" \\\n",
    "    --double-id \\\n",
    "    --pheno \"FILTERED.AMP_PD_ancestry_${ancestry}_PHENO.txt\" \\\n",
    "    --adjust \\\n",
    "    --ci 0.95 \\\n",
    "    --covar \"qc_covar_PD_unrelateds_filtered.txt\" \\\n",
    "    --covar-name SEX,AGE_BASELINE,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10 \\\n",
    "    --threads 15 \\\n",
    "    --covar-variance-standardize \\\n",
    "    --out \"${output_dir}/Logistic_FID_IID_PHENO_case_controls_${ancestry}_AMPPD_chr11_Alex\" \\\n",
    "    --glm omit-ref firth-fallback cols=+a1freq,+a1freqcc,+a1count,+totallele,+a1countcc,+totallelecc,+gcountcc,+err \\\n",
    "    --silent\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beef0b2f-075c-4838-a05f-6fda25ef1405",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "file_path = \"Logistic_FID_IID_PHENO_case_controls_AJ_AMPPD_chr11_Alex.PHENO.glm.logistic.hybrid\"\n",
    "df = pd.read_csv(file_path, sep=\"\\t\")\n",
    "\n",
    "filtered_df = df[df['TEST'] == 'ADD']\n",
    "\n",
    "filtered_df.to_csv(\"Logistic_FID_IID_PHENO_case_controls_AJ_AMPPD_chr11_Alex.PHENO.glm.logistic.hybrid.ADD_only.txt\", sep='\\t', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbb67291-4874-4068-9b94-3812d40e3280",
   "metadata": {},
   "source": [
    "## Burden analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66c350fa-af44-4896-a622-099724c2f69e",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wget https://github.com/zhanxw/rvtests/releases/download/v2.1.0/rvtests_linux64.tar.gz\n",
    "! tar -xvzf  rvtests_linux64.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "fc43c6e6-4f9c-492f-8625-f509d159618b",
   "metadata": {},
   "outputs": [],
   "source": [
    "! bgzip AMPPD_Alex_vars_splicing_exonic_EUR.vcf -k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "0fd4a785-f8ca-424f-aefc-a511e178666d",
   "metadata": {},
   "outputs": [],
   "source": [
    "! tabix -f -p vcf AMPPD_Alex_vars_splicing_exonic_EUR.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "766a80af-c519-4135-9987-5b5256ee63c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "! bgzip AMPPD_Alex_vars_splicing_exonic_AJ.vcf -k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5850120e-3e91-48c3-8e9f-8d1ef5c06935",
   "metadata": {},
   "outputs": [],
   "source": [
    "! tabix -f -p vcf AMPPD_Alex_vars_splicing_exonic_AJ.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5f79d74b-b891-400d-ab2e-2d187c102731",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "pheno = pd.read_csv(\"FILTERED.AMP_PD_ancestry_EUR_PHENO.txt\", sep=\"\\t\")\n",
    "covar = pd.read_csv(\"qc_covar_PD_unrelateds_filtered.txt\", sep=\"\\t\")\n",
    "\n",
    "merged = pd.merge(covar, pheno, on=[\"FID\", \"IID\"], how=\"inner\")\n",
    "\n",
    "merged.to_csv(\"merged_file.txt\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf0beadf-b251-40fd-9ee7-8755e82895bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "!executable/rvtest \\\n",
    "    --inVcf \"AMPPD_Alex_vars_splicing_exonic_EUR.vcf.gz\" \\\n",
    "    --out \"Alex_exonic_splicing_SORL1_Burden_EUR\" \\\n",
    "    --numThread 10 \\\n",
    "    --noweb \\\n",
    "    --hide-covar \\\n",
    "    --kernel skat,skato \\\n",
    "    --pheno \"merged_file.txt\" \\\n",
    "    --pheno-name PHENO \\\n",
    "    --geneFile \"/data/CARD_AA/data/AD_Marzieh/Alex/refFlat.txt\" \\\n",
    "    --covar \"merged_file.txt\" \\\n",
    "    --covar-name SEX,AGE_BASELINE,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10 \\\n",
    "    --multipleAllele \\\n",
    "    --gene SORL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "df455b32-0376-4629-9001-1f78d68a3003",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "pheno = pd.read_csv(\"FILTERED.AMP_PD_ancestry_AJ_PHENO.txt\", sep=\"\\t\")\n",
    "covar = pd.read_csv(\"qc_covar_PD_unrelateds_filtered.txt\", sep=\"\\t\")\n",
    "\n",
    "merged = pd.merge(covar, pheno, on=[\"FID\", \"IID\"], how=\"inner\")\n",
    "\n",
    "merged.to_csv(\"merged_file_AJ.txt\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc6725ea-c5e6-4b1a-81d0-7cf829b70225",
   "metadata": {},
   "outputs": [],
   "source": [
    "!executable/rvtest \\\n",
    "    --inVcf \"AMPPD_Alex_vars_splicing_exonic_AJ.vcf.gz\" \\\n",
    "    --out \"Alex_exonic_splicing_SORL1_Burden_AJ\" \\\n",
    "    --numThread 10 \\\n",
    "    --noweb \\\n",
    "    --hide-covar \\\n",
    "    --kernel skat,skato \\\n",
    "    --pheno \"merged_file_AJ.txt\" \\\n",
    "    --pheno-name PHENO \\\n",
    "    --geneFile \"/data/CARD_AA/data/AD_Marzieh/Alex/refFlat.txt\" \\\n",
    "    --covar \"merged_file_AJ.txt\" \\\n",
    "    --covar-name SEX,AGE_BASELINE,PC1,PC2,PC3,PC4,PC5,PC6,PC7,PC8,PC9,PC10 \\\n",
    "    --multipleAllele \\\n",
    "    --gene SORL1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b55ea1b-3699-4dc2-bdc8-a2c9e6b3e8a7",
   "metadata": {},
   "source": [
    "## Calculate of F-U and F-A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "424a94c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink/1.9\n",
    "\n",
    "vcf_file=\"/${WORK_DIR}/AMPPD_Alex_vars_splicing_exonic_EUR.vcf.gz\"\n",
    "\n",
    "ancestries=(\"EUR\")\n",
    "\n",
    "output_dir=\"/${WORK_DIR}\"\n",
    "\n",
    "mkdir -p \"$output_dir\"\n",
    "\n",
    "for ancestry in \"${ancestries[@]}\"; do\n",
    "  echo \"Processing ancestry: $ancestry\"\n",
    "\n",
    "  plink \\\n",
    "    --vcf \"$vcf_file\" \\\n",
    "    --double-id \\\n",
    "    --pheno \"FILTERED.AMP_PD_ancestry_${ancestry}_PHENO.txt\" \\\n",
    "    --pheno-name PHENO \\\n",
    "    --assoc \\\n",
    "    --adjust \\\n",
    "    --ci 0.95 \\\n",
    "    --allow-no-sex \\\n",
    "    --threads 15 \\\n",
    "    --out \"${output_dir}/Asso_FID_IID_PHENO_case_controls_${ancestry}_AMPPD_chr11_Alex\" \\\n",
    "    --silent\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d108d493-ea9e-459a-bb18-1b6b983ba95e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink/1.9\n",
    "\n",
    "vcf_file=\"/${WORK_DIR}/AMPPD_Alex_vars_splicing_exonic_AJ.vcf.gz\"\n",
    "\n",
    "ancestries=(\"AJ\")\n",
    "\n",
    "output_dir=\"/${WORK_DIR}\"\n",
    "\n",
    "mkdir -p \"$output_dir\"\n",
    "\n",
    "for ancestry in \"${ancestries[@]}\"; do\n",
    "  echo \"Processing ancestry: $ancestry\"\n",
    "\n",
    "  plink \\\n",
    "    --vcf \"$vcf_file\" \\\n",
    "    --double-id \\\n",
    "    --pheno \"FILTERED.AMP_PD_ancestry_${ancestry}_PHENO.txt\" \\\n",
    "    --pheno-name PHENO \\\n",
    "    --assoc \\\n",
    "    --adjust \\\n",
    "    --ci 0.95 \\\n",
    "    --allow-no-sex \\\n",
    "    --threads 15 \\\n",
    "    --out \"${output_dir}/Asso_FID_IID_PHENO_case_controls_${ancestry}_AMPPD_chr11_Alex\" \\\n",
    "    --silent\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python/3.10",
   "language": "python",
   "name": "py3.10"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
