{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e90a5ecf",
   "metadata": {},
   "source": [
    "# AMP DLB\n",
    "\n",
    "* **Project:** ADRD-SORL1-Biobanks\n",
    "* **Version:** Python/3.10\n",
    "* **Last Updated:** 14-Jun-2025\n",
    "\n",
    "## Notebook Overview\n",
    "\n",
    "Characterization of SORL1 variants, allele freqs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2af742af",
   "metadata": {},
   "source": [
    "# Query AMP DLB to check for variants of interest, and allele frequency"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da39a029",
   "metadata": {},
   "source": [
    "## Variables used \n",
    "- `${ANCESTRY}` = EUR, AFR, AMR, AAC, AJ, MDE, SAS, CAS, EAS, FIN, CAH\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1529a7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "82c7a11b-3ccf-4d42-8a79-97112ea07991",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>SEX</th>\n",
       "      <th>AGE_BASELINE</th>\n",
       "      <th>AGE_DIAGNOSIS</th>\n",
       "      <th>FAMILY_HISTORY</th>\n",
       "      <th>EDUCATION</th>\n",
       "      <th>LATEST_DX</th>\n",
       "      <th>DX_TYPE</th>\n",
       "      <th>ENROLL_STUDY_ARM</th>\n",
       "      <th>COHORT</th>\n",
       "      <th>PD_PHENO</th>\n",
       "      <th>PD_EXTRA_PHENO</th>\n",
       "      <th>DLB_PHENO</th>\n",
       "      <th>MSA_PHENO</th>\n",
       "      <th>PSP_PHENO</th>\n",
       "      <th>CBD_PHENO</th>\n",
       "      <th>AGE_ANALYSIS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BF-1001</td>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No PD Nor Other Neurological Disorder</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BF-1002</td>\n",
       "      <td>2</td>\n",
       "      <td>66</td>\n",
       "      <td>61.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BF-1003</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>56.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BF-1004</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BF-1005</td>\n",
       "      <td>2</td>\n",
       "      <td>61</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No PD Nor Other Neurological Disorder</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ID  SEX  AGE_BASELINE  AGE_DIAGNOSIS  FAMILY_HISTORY  EDUCATION  \\\n",
       "0  BF-1001    1            55            NaN             0.0        1.0   \n",
       "1  BF-1002    2            66           61.0             0.0        1.0   \n",
       "2  BF-1003    1            61           56.0             0.0        1.0   \n",
       "3  BF-1004    1            62           55.0             0.0        1.0   \n",
       "4  BF-1005    2            61            NaN             0.0        1.0   \n",
       "\n",
       "                               LATEST_DX  DX_TYPE ENROLL_STUDY_ARM   COHORT  \\\n",
       "0  No PD Nor Other Neurological Disorder      NaN  Healthy Control  BIOFIND   \n",
       "1                          Idiopathic PD      NaN               PD  BIOFIND   \n",
       "2                          Idiopathic PD      NaN               PD  BIOFIND   \n",
       "3                          Idiopathic PD      NaN               PD  BIOFIND   \n",
       "4  No PD Nor Other Neurological Disorder      NaN  Healthy Control  BIOFIND   \n",
       "\n",
       "   PD_PHENO  PD_EXTRA_PHENO  DLB_PHENO  MSA_PHENO  PSP_PHENO  CBD_PHENO  \\\n",
       "0         1               1          1          1          1          1   \n",
       "1         2               2         -9         -9         -9         -9   \n",
       "2         2               2         -9         -9         -9         -9   \n",
       "3         2               2         -9         -9         -9         -9   \n",
       "4         1               1          1          1          1          1   \n",
       "\n",
       "   AGE_ANALYSIS  \n",
       "0          55.0  \n",
       "1          61.0  \n",
       "2          56.0  \n",
       "3          55.0  \n",
       "4          61.0  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qc_covar_DLB2 = pd.read_csv(\"/${WORK_DIR}/addedPHENO_COVFILE_releasev3_SEPT2022_fromTerra.csv\", sep=\",\")\n",
    "qc_covar_DLB2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8582df7-57e0-4c8c-98e9-f06158757c50",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "ancestries = [\"AFR\", \"AJ\", \"EUR\", \"MDE\", \"AAC\", \"AMR\", \"FIN\", \"CAS\", \"SAS\", \"EAS\", \"CAH\"]\n",
    "\n",
    "qc_covar_DLB2 = pd.read_csv(\"/${WORK_DIR}/addedPHENO_COVFILE_releasev3_SEPT2022_fromTerra.csv\", sep=\",\")\n",
    "\n",
    "all_to_remove = set()\n",
    "\n",
    "for ancestry in ancestries:\n",
    "    remove_file = f\"/${WORK_DIR}/toRemove_1stand2ndDegree_Relateds_{ancestry}_noDups.txt\"\n",
    "    \n",
    "    try:\n",
    "        relateds = pd.read_csv(remove_file, sep=\"\\t\", header=None, names=[\"FID\", \"IID\"])\n",
    "        all_to_remove.update(relateds[\"IID\"].tolist())\n",
    "        print(f\"Loaded {len(relateds)} samples to remove from {ancestry}\")\n",
    "    except FileNotFoundError:\n",
    "        print(f\"⚠️ File not found for ancestry: {ancestry} — skipping.\")\n",
    "\n",
    "filtered_covar = qc_covar_DLB2[~qc_covar_DLB2['ID'].isin(all_to_remove)]\n",
    "\n",
    "filtered_covar.to_csv(\"qc_covar_DLB2_unrelateds_only.csv\", sep=\",\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1eb36733-d48d-477f-8e14-e4f8ea3bf159",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>SEX</th>\n",
       "      <th>AGE_BASELINE</th>\n",
       "      <th>AGE_DIAGNOSIS</th>\n",
       "      <th>FAMILY_HISTORY</th>\n",
       "      <th>EDUCATION</th>\n",
       "      <th>LATEST_DX</th>\n",
       "      <th>DX_TYPE</th>\n",
       "      <th>ENROLL_STUDY_ARM</th>\n",
       "      <th>COHORT</th>\n",
       "      <th>PD_PHENO</th>\n",
       "      <th>PD_EXTRA_PHENO</th>\n",
       "      <th>DLB_PHENO</th>\n",
       "      <th>MSA_PHENO</th>\n",
       "      <th>PSP_PHENO</th>\n",
       "      <th>CBD_PHENO</th>\n",
       "      <th>AGE_ANALYSIS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BF-1001</td>\n",
       "      <td>1</td>\n",
       "      <td>55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No PD Nor Other Neurological Disorder</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BF-1002</td>\n",
       "      <td>2</td>\n",
       "      <td>66</td>\n",
       "      <td>61.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BF-1003</td>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>56.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BF-1004</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>-9</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BF-1005</td>\n",
       "      <td>2</td>\n",
       "      <td>61</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No PD Nor Other Neurological Disorder</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>BIOFIND</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ID  SEX  AGE_BASELINE  AGE_DIAGNOSIS  FAMILY_HISTORY  EDUCATION  \\\n",
       "0  BF-1001    1            55            NaN             0.0        1.0   \n",
       "1  BF-1002    2            66           61.0             0.0        1.0   \n",
       "2  BF-1003    1            61           56.0             0.0        1.0   \n",
       "3  BF-1004    1            62           55.0             0.0        1.0   \n",
       "4  BF-1005    2            61            NaN             0.0        1.0   \n",
       "\n",
       "                               LATEST_DX  DX_TYPE ENROLL_STUDY_ARM   COHORT  \\\n",
       "0  No PD Nor Other Neurological Disorder      NaN  Healthy Control  BIOFIND   \n",
       "1                          Idiopathic PD      NaN               PD  BIOFIND   \n",
       "2                          Idiopathic PD      NaN               PD  BIOFIND   \n",
       "3                          Idiopathic PD      NaN               PD  BIOFIND   \n",
       "4  No PD Nor Other Neurological Disorder      NaN  Healthy Control  BIOFIND   \n",
       "\n",
       "   PD_PHENO  PD_EXTRA_PHENO  DLB_PHENO  MSA_PHENO  PSP_PHENO  CBD_PHENO  \\\n",
       "0         1               1          1          1          1          1   \n",
       "1         2               2         -9         -9         -9         -9   \n",
       "2         2               2         -9         -9         -9         -9   \n",
       "3         2               2         -9         -9         -9         -9   \n",
       "4         1               1          1          1          1          1   \n",
       "\n",
       "   AGE_ANALYSIS  \n",
       "0          55.0  \n",
       "1          61.0  \n",
       "2          56.0  \n",
       "3          55.0  \n",
       "4          61.0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qc_covar_DLB2 = pd.read_csv(\"qc_covar_DLB2_unrelateds_only.csv\", sep=\",\")\n",
    "qc_covar_DLB2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e916b582-3fef-43ce-8648-eb81603e5cbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_case_DLB2 = qc_covar_DLB2[qc_covar_DLB2[\"DLB_PHENO\"]==2]\n",
    "qc_case_DLB2.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c3ab0be6-c6b1-427f-b3ec-a5644221b9ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_case_DLB2_plink = qc_case_DLB2[[\"ID\"]]\n",
    "qc_case_DLB2_plink.to_csv(\"qc_case_DLB2_plink.txt\", sep=\",\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "711f5203-ce91-44e5-8f1d-acee8e536ee0",
   "metadata": {},
   "outputs": [],
   "source": [
    "!awk '{print $0, $1}' qc_case_DLB2_plink.txt > qc_case_DLB2_ID_ID_plink.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c048f0a3-e8c7-4d22-9451-570f40d7bf05",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_control_DLB2 = qc_covar_DLB2[qc_covar_DLB2[\"PD_PHENO\"]==1]\n",
    "qc_control_DLB2.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e6a0ad28-46b2-4210-838b-081a35953490",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_control_DLB2_plink = qc_control_DLB2[[\"ID\"]]\n",
    "qc_control_DLB2_plink.to_csv(\"qc_control_DLB2_plink.txt\", sep=\",\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "03ef0666-2e63-479a-8400-063075bca319",
   "metadata": {},
   "outputs": [],
   "source": [
    "!awk '{print $0, $1}' qc_control_DLB2_plink.txt > qc_control_DLB2_ID_ID_plink.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6561929-2abe-4596-acd5-1178927f8d85",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink\n",
    "plink2 --pfile /${WORK_DIR}/FILTERED.AMP_PD_${ANCESTRY} \\\n",
    "--chr 11 --from-bp 121452314 --to-bp 121633763  --make-bed --out AMPDLB2_Alex_vars_all_${ANCESTRY}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1148cb8-2f37-45e7-9415-1ff1021799cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink/1.9\n",
    "\n",
    "plink --bfile AMPDLB2_Alex_vars_all_${ANCESTRY} --keep qc_case_DLB2_ID_ID_plink.txt --make-bed --out AMPDLB2_Alex_vars_all_${ANCESTRY}_cases2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80907705-bebe-4788-8c5b-eb641e11ccd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink/1.9\n",
    "\n",
    "plink --bfile AMPDLB2_Alex_vars_all_${ANCESTRY}_cases2 --freq --out AMPDLB2_Alex_vars_all_${ANCESTRY}_cases2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4fe9ad8-ab34-41c5-933f-00f0bb68f22e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink/1.9\n",
    "\n",
    "plink --bfile AMPDLB2_Alex_vars_all_${ANCESTRY} --keep qc_control_DLB2_ID_ID_plink.txt --make-bed --out AMPDLB2_Alex_vars_all_${ANCESTRY}_controls2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "596ef0df-11f7-4060-aef0-04cd135d10ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink/1.9\n",
    "\n",
    "plink --bfile AMPDLB2_Alex_vars_all_${ANCESTRY}_controls2 --freq --out AMPDLB2_Alex_vars_all_${ANCESTRY}_controls2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78afd65a-ecde-460e-a8ec-43c62d0f2294",
   "metadata": {},
   "source": [
    "## Check Zygosity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33b4df17-d414-43ce-be98-8a31d16c401c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink\n",
    "plink2 --bfile AMPDLB2_Alex_vars_all_${ANCESTRY} --recode A --out AMPDLB2_Alex_vars_all_${ANCESTRY}_recode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35c27bc8-47e2-4f09-8ba9-1e7c022778ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink\n",
    "plink2 --bfile AMPDLB2_Alex_vars_all_${ANCESTRY}_cases2 --recode A --out AMPDLB2_Alex_vars_all_${ANCESTRY}_cases2_recode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e38bc48-13f1-409e-8a33-c0d199749f37",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load plink\n",
    "plink2 --bfile AMPDLB2_Alex_vars_all_${ANCESTRY}_controls2 --recode A --out AMPDLB2_Alex_vars_all_${ANCESTRY}_controls2_recode"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2672647a-3c60-4915-9581-63c3d6b08f2d",
   "metadata": {},
   "source": [
    "## Merge files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "947ba449-7589-4ee2-a399-4620361bc734",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "control_amp_paths = glob.glob(\"AMPDLB2_*controls2*.frq\")\n",
    "case_amp_paths = glob.glob(\"AMPDLB2_*cases2*.frq\")\n",
    "\n",
    "control_amp_paths.sort()\n",
    "case_amp_paths.sort()\n",
    "\n",
    "amp_paths = list(zip(control_amp_paths, case_amp_paths))\n",
    "print(f\"Total paired files: {len(amp_paths)}\")\n",
    "\n",
    "df_list = []\n",
    "for paths in amp_paths:\n",
    "    for i, path in enumerate(paths):\n",
    "        df = pd.read_csv(path, sep=r'\\s+', engine='python')\n",
    "\n",
    "        df[\"NCHROBS\"] = pd.to_numeric(df[\"NCHROBS\"])\n",
    "        nchrobs = df[\"NCHROBS\"].max()\n",
    "        ancestry = path.split(\"_\")[4]\n",
    "        maf_col_name = f\"{ancestry} {'Controls' if i == 0 else 'Cases'} MAF (NCHROBS = {nchrobs})\"\n",
    "        df.rename({\"MAF\": maf_col_name}, inplace=True, axis=1)\n",
    "\n",
    "        df = df[[\"SNP\", maf_col_name]]\n",
    "        display(df.head(3))\n",
    "\n",
    "        df_list.append(df)\n",
    "\n",
    "for i in range(len(df_list)):\n",
    "    print(f\"Match with first file [{i}]:\", list(df_list[0][\"SNP\"]) == list(df_list[i][\"SNP\"]))\n",
    "\n",
    "merged_df = df_list[0]\n",
    "for df in df_list[1:]:\n",
    "    merged_df = pd.merge(merged_df, df, on='SNP', how=\"outer\")\n",
    "\n",
    "merged_df = merged_df.fillna(0)\n",
    "\n",
    "display(merged_df.head())\n",
    "\n",
    "os.makedirs(\"Merged_AMPDLB2\", exist_ok=True)\n",
    "merged_df.to_csv(\"Merged_AMPDLB2/AMPDLB2.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12d899b5",
   "metadata": {},
   "source": [
    "## Count the number of cases and controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "38be3b7f-419e-4773-ae64-470f4806864b",
   "metadata": {},
   "outputs": [],
   "source": [
    "! awk 'NR==FNR {ids[$1]; next} $1 in ids' qc_case_DLB2_ID_ID_plink.txt /${WORK_DIR}/FILTERED.AMP_PD_ancestry_${ANCESTRY}.samples > filtered_DLB2_samples2_${ANCESTRY}.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53720945-44da-4e51-8610-2cc20e2ba9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc filtered_DLB2_samples2_${ANCESTRY}.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "222abc4d-c37a-4c44-91d6-9a57a2c46b0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "! awk 'NR==FNR {ids[$1]; next} $1 in ids' qc_control_DLB2_ID_ID_plink.txt /${WORK_DIR}/FILTERED.AMP_PD_ancestry_${ANCESTRY}.samples > filtered_samples__DLB2_controls2_${ANCESTRY}.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f3ead21-7ad5-4a07-99db-dd9f6eda0d2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc filtered_samples__DLB2_controls2_${ANCESTRY}.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cd816bd-5876-4608-a403-190657543eaa",
   "metadata": {},
   "source": [
    "## Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17874826",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load annovar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c60b9823",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "table_annovar.pl AMP_DLB_8.cleaned.vcf $ANNOVAR_DATA/hg38 \\\n",
    "    --buildver hg38 \\\n",
    "    --remove \\\n",
    "    --thread 48 \\\n",
    "    --maxgenethread 48 \\\n",
    "    --protocol refGene,clinvar_20140902,avsnp151,dbnsfp47a \\\n",
    "    --operation g,f,f,f \\\n",
    "    --nopolish \\\n",
    "    --nastring . \\\n",
    "    --out AMP_DLB_8.vcf.anno \\\n",
    "    --vcfinput"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8f14d9c",
   "metadata": {},
   "source": [
    "## Merge files and apply criteria (exonic and splicing variants, mac 2, CADD>20, and Only present in cases)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "c27be8fb-54db-41aa-92e3-e7fd2f6f38a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"AMP_DLB_8.vcf.anno.hg38_multianno.txt\", sep=\"\\t\")\n",
    "\n",
    "columns_to_keep = [\n",
    "    \"Chr\", \"Start\", \"End\", \"Ref\", \"Alt\",\n",
    "    \"Func.refGene\", \"Gene.refGene\", \"GeneDetail.refGene\",\n",
    "    \"ExonicFunc.refGene\", \"AAChange.refGene\",\n",
    "    \"clinvar_20140902\", \"avsnp151\", \"CADD_phred\"\n",
    "]\n",
    "\n",
    "filtered_df = df[columns_to_keep]\n",
    "\n",
    "filtered_df.to_csv(\"AMP_DLB_8.filtered.txt\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "2cde8bd6-0b2b-4792-b321-78afa43baa78",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "anno = pd.read_csv('AMP_DLB_8.filtered.txt', sep='\\t')\n",
    "freq = pd.read_csv('Merged_AMPDLB2/AMPDLB2.csv')\n",
    "\n",
    "freq['Start'] = freq['SNP'].apply(lambda x: int(x.split(':')[1]))\n",
    "\n",
    "merged = pd.merge(anno, freq, on='Start', how='inner')\n",
    "\n",
    "merged.to_csv('AMP_DLB2_8_merged_output2.txt', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8aa10ff-65f5-483b-8f82-263d6928664a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "ancestries = [\"AFR\",\"AJ\",\"EUR\",\"MDE\",\"AAC\",\"AMR\",\"FIN\",\"CAS\",\"SAS\",\"EAS\",\"CAH\"]\n",
    "zyg_list = []\n",
    "\n",
    "for ancestry in ancestries:\n",
    "    for group in [\"cases\", \"controls\"]:\n",
    "        filename = f\"AMPDLB2_Alex_vars_all_{ancestry}_{group}2_recode.raw\" \n",
    "        if os.path.exists(filename):\n",
    "            print(f\"Processing: {filename}\")\n",
    "            df = pd.read_csv(filename, sep=\"\\t\")\n",
    "            df.drop(columns=[\"FID\", \"IID\", \"PAT\", \"MAT\", \"SEX\", \"PHENOTYPE\"], inplace=True)\n",
    "            df = df.fillna(2).astype(int)\n",
    "\n",
    "            snps = [snpid.split(\"_\")[0] for snpid in df.columns]\n",
    "            hom_alt = (df == 0).sum()\n",
    "            het = (df == 1).sum()\n",
    "\n",
    "            df_zyg = pd.DataFrame({\n",
    "                \"SNP\": snps,\n",
    "                f\"{ancestry}_{group}_hom_alt_ac\": hom_alt.values,\n",
    "                f\"{ancestry}_{group}_het_ac\": het.values\n",
    "            })\n",
    "\n",
    "            zyg_list.append(df_zyg)\n",
    "        else:\n",
    "            print(f\"Missing file: {filename}\")\n",
    "\n",
    "from functools import reduce\n",
    "\n",
    "if zyg_list:\n",
    "    merged_zyg = reduce(lambda left, right: pd.merge(left, right, on='SNP', how='outer'), zyg_list)\n",
    "\n",
    "    merged = pd.read_csv(\"AMP_DLB2_8_merged_output2.txt\", sep=\"\\t\")  \n",
    "    final_merged = pd.merge(merged, merged_zyg, on='SNP', how='left')\n",
    "    final_merged.to_csv(\"AMP_DLB2_8_final_output2.txt\", sep=\"\\t\", index=False)\n",
    "    print(\"Final merged file written: AMP_DLB2_8_final_output2.txt\")\n",
    "else:\n",
    "    print(\"No valid zygosity data found to merge.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e962fe70-2ab3-4b51-9a70-e3fa53481a89",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_amp_merged = pd.read_csv(\"AMP_DLB2_8_final_output2.txt\", sep=\"\\t\")\n",
    "\n",
    "amp_freq_cols = df_amp_merged.filter(like=\"NCHROBS\").columns\n",
    "\n",
    "criteria_amp_anno = (\n",
    "    (df_amp_merged[\"Func.refGene\"].isin([\"splicing\", \"exonic\"])) &\n",
    "    ~(df_amp_merged[\"ExonicFunc.refGene\"] == \"synonymous SNV\")\n",
    ")\n",
    "\n",
    "df_amp_filtered = df_amp_merged[criteria_amp_anno]\n",
    "\n",
    "criteria_amp_freq = (df_amp_filtered[amp_freq_cols] > 0.0).any(axis=1)\n",
    "df_amp_filtered = df_amp_filtered[criteria_amp_freq]\n",
    "\n",
    "df_amp_filtered.to_csv(\"AMP_DLB2_8_filtered_coding_output2.txt\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2aad1a64-d407-4a61-ae25-1b80b9c7f0b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc AMP_DLB2_8_filtered_coding_output2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40a7136f-2e94-48aa-9839-9e4ab2feb0db",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_amp_filtered = pd.read_csv(\"AMP_DLB2_8_filtered_coding_output2.txt\", sep=\"\\t\")\n",
    "\n",
    "amp_controls_cols = df_amp_filtered.filter(like=\"Controls MAF\").columns\n",
    "\n",
    "df_amp_filtered_casesonly = df_amp_filtered[(df_amp_filtered[amp_controls_cols] == 0.0).all(axis=1)]\n",
    "\n",
    "df_amp_filtered_casesonly.to_csv(\"AMP_DLB2_8__coding_casesonly_output2.txt\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c83d3dba-1a51-44fc-abba-4b688fde1463",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc AMP_DLB2_8__coding_casesonly_output2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5662fab-b03b-4d0d-9dc9-82d50ac3c2f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_amp_casesonly = pd.read_csv(\"AMP_DLB2_8__coding_casesonly_output2.txt\", sep=\"\\t\")\n",
    "\n",
    "criteria_amp_cadd = (df_amp_casesonly[\"CADD_phred\"] >= 20) | (df_amp_casesonly[\"CADD_phred\"].isna())\n",
    "df_amp_casesonly_cadd_filtered = df_amp_casesonly[criteria_amp_cadd]\n",
    "\n",
    "df_amp_casesonly_cadd_filtered.to_csv(\"AMP_DLB2_8__coding_casesonly_CADD20_output2.txt\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecafc052-5f7a-4ec1-90b1-9a79da58ab58",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc AMP_DLB2_8__coding_casesonly_CADD20_output2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e880d00-d1e1-4146-a579-2f493e13216e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_amp_casesonly_cadd = pd.read_csv(\"AMP_DLB2_8__coding_casesonly_CADD20_output2.txt\", sep=\"\\t\")\n",
    "\n",
    "criteria_${ANCESTRY}_ac = (df_amp_casesonly_cadd[\"${ANCESTRY}_cases_het_ac\"] >= 2) | (df_amp_casesonly_cadd[\"${ANCESTRY}_cases_hom_alt_ac\"] >= 2)\n",
    "df_amp_casesonly_cadd_${ANCESTRY}_filtered = df_amp_casesonly_cadd[criteria_${ANCESTRY}_ac]\n",
    "\n",
    "df_amp_casesonly_cadd_${ANCESTRY}_filtered.to_csv(\"AMP_DLB2_8__coding_casesonly_CADD20_${ANCESTRY}AC2_output2.txt\", sep=\"\\t\", index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a57431e-4337-422e-99a1-4a1a1b882ed6",
   "metadata": {},
   "outputs": [],
   "source": [
    "! wc AMP_DLB2_8__coding_casesonly_CADD20_${ANCESTRY}AC2_output2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ff6522a-0996-46f8-9e6c-dad1aa7fe2f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "! cat AMP_DLB2_8__coding_casesonly_CADD20_${ANCESTRY}AC2_output2.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47f72459-3147-4d0a-870c-54b92a1158d8",
   "metadata": {},
   "source": [
    "## Caunt total variansts per ancestry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54851939-f51f-4f17-9051-9bc1ac4ee47b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"AMP_DLB2_8_merged_output2.txt\", sep=\"\\t\")\n",
    "\n",
    "ancestries = [\"AJ\", \"EUR\", \"FIN\", \"MDE\"]\n",
    "\n",
    "summary_tables = []\n",
    "\n",
    "for ancestry in ancestries:\n",
    "    control_col = [col for col in df.columns if col.startswith(f\"{ancestry} Controls MAF\")][0]\n",
    "    case_col = [col for col in df.columns if col.startswith(f\"{ancestry} Cases MAF\")][0]\n",
    "\n",
    "    ancestry_variants = df[(df[control_col] > 0) | (df[case_col] > 0)].copy()\n",
    "\n",
    "    if not ancestry_variants.empty:\n",
    "        \n",
    "        func_counts = ancestry_variants.groupby(\"Func.refGene\").size().reset_index(name=\"VariantCount\")\n",
    "        func_counts[\"Ancestry\"] = ancestry\n",
    "\n",
    "        exonic_variants = ancestry_variants[ancestry_variants[\"Func.refGene\"] == \"exonic\"]\n",
    "        exonic_counts = exonic_variants.groupby(\"ExonicFunc.refGene\").size().reset_index(name=\"VariantCount\")\n",
    "        exonic_counts[\"Ancestry\"] = ancestry\n",
    "        exonic_counts[\"Func.refGene\"] = \"exonic\"\n",
    "\n",
    "        summary_tables.append(func_counts)\n",
    "        summary_tables.append(exonic_counts)\n",
    "\n",
    "final_summary = pd.concat(summary_tables, ignore_index=True)\n",
    "\n",
    "final_summary.to_csv(\"AMP_DLB2_variant_ancestry_func_and_exonic_type_counts2.txt\", sep=\"\\t\", index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "514c0a93-9fee-481c-819c-f85982048364",
   "metadata": {},
   "outputs": [],
   "source": [
    "! cat AMP_DLB2_variant_ancestry_func_and_exonic_type_counts2.txt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python/3.10",
   "language": "python",
   "name": "py3.10"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
